// @ux/tabs/src/styles.js

import { css } from 'linaria';
import intents from './intents.config.js';

import { mixins } from '@ux/stylevars';

const {
  rtlIcon,
  mediaBreakpointDown
} = mixins;

export default css`

:global() {
  @import "~@ux/icon/chevron-left/index.css";
  @import "~@ux/icon/chevron-right/index.css";

  .ux-conveyor-clip {
    position: relative;
    overflow: hidden;
  }

  .ux-conveyor {
    position: relative;
    overflow-x: auto;

    .ux-tabs {
      border-bottom: 1px solid ${intents.uxTabs.borderColor};
      position: relative;
      transition: margin-left .5s;
      display: flex;
      flex-wrap: nowrap;
      padding: 0;
      margin-bottom: 0;
    }

    .ux-tab-item {
      display: inline-flex;
      margin-top: 0;
      margin-bottom: -1px;
    }

    .ux-tabs .ux-tab-link,
    .ux-tabs .ux-tab-link-underline {
      display: flex;
      align-items: center;
      justify-content: center;
      background-color: transparent;
      color: ${intents.uxTabs.item.foregroundColor};
      border: 1px solid transparent;
      border-bottom-color: ${intents.uxTabs.borderColor};
      border-radius: 0;
      border-top-left-radius: ${intents.uxTabs.borderRadius};
      border-top-right-radius: ${intents.uxTabs.borderRadius};
      font-family: ${intents.uxTabs.item.fontFamily};
      font-weight: ${intents.uxTabs.item.fontWeight};
      font-size: ${intents.uxTabs.item.fontSize};
      line-height: ${intents.uxTabs.item.lineHeight};
      cursor: pointer;
      padding: calc(${intents.ux.box.density} * 3) calc(${intents.ux.box.density} * 6);
      white-space: nowrap;
      position: relative;
      &:focus {
        outline: none;
      }

      &:focus:before {
        content: '';
        position: absolute;
        left: 0;
        right: 0;
        top: 0;
        bottom: 0;
        width: 100%;
        border-left: var(--uxp-focus-visible-outline-width) var(--uxp-focus-visible-outline-style) var(--uxp-focus-visible-outline-color);
        border-right: var(--uxp-focus-visible-outline-width) var(--uxp-focus-visible-outline-style) var(--uxp-focus-visible-outline-color);
        border-top: var(--uxp-focus-visible-outline-width) var(--uxp-focus-visible-outline-style) var(--uxp-focus-visible-outline-color);
      }

      ${mediaBreakpointDown('sm', `
        padding: calc(${intents.ux.box.density} * 2) calc(${intents.ux.box.density} * 4);
      `)}
    }

    .ux-tabs .ux-tab-link:hover,
    .ux-tabs .ux-tab-link:focus,
    .ux-tabs .ux-tab-link-underline:hover,
    .ux-tabs .ux-tab-link-underline:focus{
      border: 1px solid transparent;
      border-bottom-color: ${intents.uxTabs.borderColor};
      color: ${intents.uxTabs.itemHovered.foregroundColor};
    }

    .ux-tabs .ux-tab-link.active,
    .ux-tabs .ux-tab-link.active:hover,
    .ux-tabs .ux-tab-link.active:focus {
      border: 1px solid ${intents.uxTabs.borderColor};
      color: ${intents.uxTabs.foregroundColor};
      border-bottom-color: transparent;
      background-color: ${intents.uxTabs.backgroundColor};
      font-weight: bolder;
    }

    .ux-tabs .ux-tab-link-underline.active,
    .ux-tabs .ux-tab-link-underline.active:hover,
    .ux-tabs .ux-tab-link-underline.active:focus {
      border: 1px solid transparent;
      color: ${intents.uxTabs.foregroundColor};
      border-bottom-color: ${intents.uxTabs.itemChosen.borderColor};
      border-bottom-width: 3px;
      background-color: ${intents.uxTabs.backgroundColor};
      font-weight: bolder;

      &:after {
        display: none;
      }
    }
  }

  .ux-tab-content.ux-tab-content-hidden {
    display: none;
  }

  .ux-conveyor-control {
    position: absolute;
    top: 0;
    bottom: 0;
    width: 2.5rem;
    background-color: ${intents.uxTabs.backgroundColor};
    color: ${intents.uxTabs.item.foregroundColor};
    border: 1px solid ${intents.uxTabs.borderColor};
    border-radius: ${intents.uxTabs.borderRadius};
    text-decoration: none;
    align-items: center;
    justify-content: center;
    cursor: pointer;

    &:hover,
    &:focus {
      outline: 0;
      color: ${intents.uxTabs.itemHovered.foregroundColor};
    }

    svg {
      ${rtlIcon()}
      width: 1rem;
    }

    &.ux-conveyor-control-left {
      left: 0;
    }

    &.ux-conveyor-control-right {
      right: 0;
    }
  }
}
`;
