// @ux/style-common/src/app.js
import { css } from 'linaria';
import { mediaBreakpointUp } from './mixins/breakpoints';
import intents from './intents.config.js';

export default css`
:global() {
  .ux-app {
    background-color: ${intents.ux.boxLowContrast.backgroundColor}; /* product faint */
    color: ${intents.ux.boxLowContrast.foregroundColor}; /* gray midrange */
  }

  .ux-pagetitle {
    padding: calc(${intents.ux.box.density} * 12.25) calc(${intents.ux.box.density} * 2);
    margin-bottom: calc(${intents.ux.box.density} * 8);
    background-color: ${intents.ux.box.backgroundColor};
    box-shadow: 0 4px 0 rgba(0, 0, 0, .1);
    ${mediaBreakpointUp('md', `
      padding-right: 0;
      padding-left: 0;
    `)}
    h1 {
      margin: 0;
      + p {
        margin: calc(${intents.ux.box.density} * 1) 0 0;
        font-size: ${intents.ux.textParagraph.fontSize}; /* 1rem */
      }
    }
    .breadcrumb {
      margin: 0 0 calc(${intents.ux.box.density} * 2);
    }
  }

  .ux-pagetitle-show-icn {
    position: relative;
    ${mediaBreakpointUp('md', `padding-left: 100px;`)}
    .ux-pagetitle-icn {
      font-size: 7rem;
      color: ${intents.ux.box.foregroundColor}; /* black dark */
      position: absolute;
      top: 0;
      left: 0;
      display: none;
      ${mediaBreakpointUp('md', `display: block;`)}
    }
  }
}
`;
