// @ux/style-common/src/card.js
import { css } from 'linaria';
import intents from './intents.config.js';
import { mediaBreakpointUp, mediaBreakpointDown } from './mixins/breakpoints';

export default css`
/* stylelint-disable max-nesting-depth */
:global() {
  .card {
    border-radius: ${intents.ux.box.borderRadius};
    background-color: ${intents.ux.box.backgroundColor}; /* white */
    border: 1px solid ${intents.ux.box.borderColor}; /* product */
    position: relative;
    display: flex;
    flex-direction: column;
    min-width: 0;
    word-wrap: break-word;
    background-clip: border-box;
    margin-bottom: calc(${intents.ux.box.density} * 5);

    > hr {
      margin-right: 0;
      margin-left: 0;
    }

    > .list-group:first-child {
      .list-group-item:first-child {
        border-radius: ${intents.ux.box.borderRadius} ${intents.ux.box.borderRadius} 0 0;
      }
    }

    > .list-group:last-child {
      .list-group-item:last-child {
        border-radius: 0 0 ${intents.ux.box.borderRadius} ${intents.ux.box.borderRadius};
      }
    }
  }

  /* Needed for old cards */
  .card-block {
    padding: calc(${intents.ux.box.density} * 5);
  }

  .card-body {
    /* Enable 'flex-grow: 1' for decks and groups so that card blocks take up
      as much space as possible, ensuring footers are aligned to the bottom. */
    flex: 1 1 auto;
    padding: calc(${intents.ux.box.density} * 5);
  }

  .card-title {
    margin-bottom: calc(${intents.ux.box.density} * 3);
    text-align: center;
    text-transform: none;
  }

  .card-subtitle {
    margin-top: calc(${intents.ux.box.density} * -1.5);
    margin-bottom: 0;
  }

  .card-text:last-child {
    margin-bottom: 0;
  }

  .card-link {
    &:hover {
      text-decoration: none;
    }

    + .card-link {
      margin-left: calc(${intents.ux.box.density} * 5);
    }
  }

  /* Optional textual caps */

  .card-header {
    background-color: ${intents.ux.boxLowContrast.backgroundColor};
    border-bottom: 1px solid ${intents.ux.box.borderColor};
    padding: calc(${intents.ux.box.density} * 3) calc(${intents.ux.box.density} * 5);
    margin-bottom: 0; /* Removes the default margin-bottom of <hN> */

    &:first-child {
      border-radius: ${intents.ux.box.borderRadius} ${intents.ux.box.borderRadius} 0 0;
    }

    + .list-group {
      .list-group-item:first-child {
        border-top: 0;
      }
    }
  }

  .card-footer {
    padding: calc(${intents.ux.box.density} * 3) calc(${intents.ux.box.density} * 5);
    background-color: rgba(0, 0, 0, .03);
    border-top: 1px solid ${intents.ux.box.borderColor};

    &:last-child {
      border-radius: 0 0 ${intents.ux.box.borderRadius} ${intents.ux.box.borderRadius};
    }
  }



  /* Header navs */

  .card-header-tabs {
    margin-right: calc(${intents.ux.box.density} * -2.5);
    margin-bottom: calc(${intents.ux.box.density} * -3);
    margin-left: calc(${intents.ux.box.density} * -2.5);
    border-bottom: 0;
  }

  .card-header-pills {
    margin-right: calc(${intents.ux.box.density} * -2.5);
    margin-left: calc(${intents.ux.box.density} * -2.5);
  }

  /* Card image */
  .card-img-overlay {
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    padding: calc(${intents.ux.box.density} * 5);
  }

  .card-img {
    width: 100%; /* Required because we use flexbox and this inherently applies align-self: stretch */
    border-radius: ${intents.ux.box.borderRadius};
  }

  /* Card image caps */
  .card-img-top {
    width: 100%; /* Required because we use flexbox and this inherently applies align-self: stretch */
    border-radius: ${intents.ux.box.borderRadius} ${intents.ux.box.borderRadius} 0 0;
  }

  .card-img-bottom {
    width: 100%; /* Required because we use flexbox and this inherently applies align-self: stretch */
    border-radius: 0 0 ${intents.ux.box.borderRadius} ${intents.ux.box.borderRadius};
  }


  /* Card deck */

  .card-deck {
    display: flex;
    flex-direction: column;

    .card {
      margin-bottom: calc(${intents.ux.box.density} * 2.5);
    }

    ${mediaBreakpointUp('sm', `
      flex-flow: row wrap;
      margin-right: calc(${intents.ux.box.density} * -2.5);
      margin-left: calc(${intents.ux.box.density} * -2.5);

      .card {
        display: flex;
        /* Flexbugs #4: https://github.com/philipwalton/flexbugs#flexbug-4 */
        flex: 1 0 0%;
        flex-direction: column;
        margin-right: calc(${intents.ux.box.density} * 2.5);
        margin-bottom: 0; /* Override the default */
        margin-left: calc(${intents.ux.box.density} * 2.5);
      }
    `)}
  }


  /* Card groups */

  .card-group {
    display: flex;
    flex-direction: column;

    /* The child selector allows nested '.card' within '.card-group' to display properly. */
    > .card {
      margin-bottom: calc(${intents.ux.box.density} * 2);
    }

    ${mediaBreakpointUp('sm', `
      flex-flow: row wrap;
      /* The child selector allows nested '.card' within '.card-group' to display properly. */
      > .card {
        /* Flexbugs #4: https://github.com/philipwalton/flexbugs#flexbug-4 */
        flex: 1 0 0%;
        margin-bottom: 0;

        + .card {
          margin-left: 0;
          border-left: 0;
        }
      }
    `)}
  }

  /* Columns */

  .card-columns {
    .card {
      margin-bottom: calc(${intents.ux.box.density} * 3);
    }

    ${mediaBreakpointUp('sm', `
      column-count: 3;
      column-gap: calc(${intents.ux.box.density} * 5);
      orphans: 1;
      widows: 1;

      .card {
        display: inline-block; /* Don't let them vertically span multiple columns */
        width: 100%; /* Don't let their width change */
      }
    `)}
  }


  /* Accordion */

  .accordion {
    .card:not(:first-of-type):not(:last-of-type) {
      border-bottom: 0;
      border-radius: 0;
    }

    .card:not(:first-of-type) {
      .card-header:first-child {
        border-radius: 0;
      }
    }

    .card:first-of-type {
      border-bottom: 0;
      border-bottom-right-radius: 0;
      border-bottom-left-radius: 0;
    }

    .card:last-of-type {
      border-top-left-radius: 0;
      border-top-right-radius: 0;
    }
  }

  /* UX Card */
  .ux-card {
    border-radius: ${intents.ux.box.borderRadius};
    margin: 0 0 calc(${intents.ux.box.density} * 5);
    color: ${intents.ux.box.foregroundColor}; /* black */
    border: 1px solid ${intents.ux.box.borderColor}; /* product */

    ${mediaBreakpointUp('xl', `margin-bottom: calc(${intents.ux.box.density} * 10);`)}

    .card-block {
      padding: calc(${intents.ux.box.density} * 5);
    }

    .ux-card {
      box-shadow: none;
    }
  }

  /* Brings card to edge on mobile
   (only xs col has padding on mobile) */
  [class*="col-xs"] > .ux-card {
    ${mediaBreakpointDown('sm', `
      margin-left: calc(${intents.ux.box.density} * -2);
      margin-right: calc(${intents.ux.box.density} * -2);
    `)}
  }
}
`;
