// @ux/style-common/src/close.js
import { css } from 'linaria';
import intents from './intents.config.js';

export default css`
:global() {
  .close {
    font-family: ${intents.ux.text.fontFamily};
    font-size: ${intents.ux.textAction.fontSize};
    font-weight: ${intents.ux.textAction.fontWeight};
    color: ${intents.ux.action.foregroundColor};
    float: right;
    line-height: 1;
    text-shadow: none;
    opacity: 1;
    width: 1em; /* Increase target area */

    &:before {
      content: '×'; /* We need to control content */
    }

    &:not(:disabled):not(.disabled) {
      &:focus,
      &:hover {
        color: ${intents.ux.action.foregroundColor};
        text-decoration: none;
        opacity: 1;
        outline: none;
      }
      /* Opinionated: add "hand" cursor to non-disabled .close elements */
      cursor: pointer;
    }

    &:focus,
    &:hover {
      opacity: 1;
      outline: none;
    }

    span {
      display: none; /* We need to control content */
    }
  }

  /* Additional properties for button version
    iOS requires the button element instead of an anchor tag.
    If you want the anchor version, it requires 'href="#"'.
    See https://developer.mozilla.org/en-US/docs/Web/Events/click#Safari_Mobile
  */

  /* stylelint-disable property-no-vendor-prefix, selector-no-qualifying-type */
  button.close {
    padding: 0;
    background-color: transparent;
    border: 0;
    -webkit-appearance: none;
  }
  /* stylelint-enable */
}
`;
