// @ux/style-common/src/images.js
import { css } from 'linaria';
import intents from './intents.config.js';

export default css`
:global() {
  /* Responsive images (ensure images don't scale beyond their parents)
  *
  * This is purposefully opt-in via an explicit class rather than being the default for all '<img>'s.
  *  We previously tried the "images are responsive by default" approach in Bootstrap v2,
  *  and abandoned it in Bootstrap v3 because it breaks lots of third-party widgets (including Google Maps)
  * which weren't expecting the images within themselves to be involuntarily resized.
  * See also https://github.com/twbs/bootstrap/issues/18178
  */
  .img-fluid {
    max-width: 100%;
    height: auto;
  }

  /* Rounded corners */
  .img-rounded {
    border-radius: calc(${intents.ux.box.borderRadius} + .125rem);
  }

  /* Image thumbnails */
  .img-thumbnail {
    background-color: ${intents.ux.box.backgroundColor};
    border: 1px solid ${intents.ux.box.borderColor};
    border-radius: ${intents.ux.box.borderRadius};
    padding: calc(${intents.ux.box.density} * 1);
    transition: all .2s ease-in-out;
    box-shadow: 0 1px 2px rgba(0, 0, 0, .075);

    /* Keep them at most 100% wide */
    max-width: 100%;
    height: auto;
  }

  /* Figures */

  .figure {
    /* Ensures the caption's text aligns with the image. */
    display: inline-block;
  }

  .figure-img {
    margin-bottom: calc(${intents.ux.box.density} * 2);
    line-height: 1;
  }

  .figure-caption {
    font-size: ${intents.ux.textCaption.fontSize};
    color: ${intents.ux.box.foregroundColor};
    opacity: .8;
  }
}
`;
