// @ux/style-common/src/input-group.js
import { css } from 'linaria';

export default css`
:global() {
  /* stylelint-disable selector-no-qualifying-type */

  /* Base styles */

  .input-group {
    position: relative;
    display: flex;
    align-items: stretch;
    width: 100%;
    z-index: 1;

    label { /* this is to hide the label that is breaking input-groups */
      border: 0;
      clip: rect(0 0 0 0);
      height: 1px;
      margin: -1px;
      overflow: hidden;
      padding: 0;
      position: absolute;
      width: 1px;
    }

    .form-control {
      /* Ensure that the input is always above the *appended* addon button for
        proper border colors. */
      position: relative;
      z-index: 2;
      flex: 1 1 auto;
      margin-bottom: 0;

      /* Bring the "active" form control to the front */
      &:focus,
      &:active,
      &:hover {
        z-index: 4;
      }
    }
  }

  .input-group-btn,
  .input-group .form-control {
    display: flex;
    align-items: center;
    &:not(:first-child):not(:last-child) {
      border-radius: 0;
    }
  }

  .input-group-btn {
    white-space: nowrap;
  }

    /* Reset rounded corners */

  .input-group .form-control:not(:last-child),
  .input-group-btn:not(:last-child) > .btn,
  .input-group-btn:not(:last-child) > .btn-group > .btn,
  .input-group-btn:not(:last-child) > .dropdown-toggle,
  .input-group-btn:not(:first-child) > .btn:not(:last-child):not(.dropdown-toggle),
  .input-group-btn:not(:first-child) > .btn-group:not(:last-child) > .btn {
    border-bottom-right-radius: 0;
    border-top-right-radius: 0;
  }

  .input-group .form-control:not(:first-child),
  .input-group-btn:not(:first-child) > .btn,
  .input-group-btn:not(:first-child) > .btn-group > .btn,
  .input-group-btn:not(:first-child) > .dropdown-toggle,
  .input-group-btn:not(:last-child) > .btn:not(:first-child),
  .input-group-btn:not(:last-child) > .btn-group:not(:first-child) > .btn {
    border-bottom-left-radius: 0;
    border-top-left-radius: 0;
  }

  /* Button input groups */

  .input-group-btn {
    position: relative;
    align-items: stretch;
    z-index: 3;
    flex-shrink: 0;
    /* Jankily prevent input button groups from wrapping with 'white-space' and
       'font-size' in combination with 'inline-block' on buttons. */
    font-size: 0;
    white-space: nowrap;

    /* Negative margin for spacing, position for bringing hovered/focused/actived
       element above the siblings. */
    > .btn {
      position: relative;

      + .btn {
        margin-left: -1px;
      }

      /* Bring the "active" button to the front */
      &:focus,
      &:active,
      &:hover {
        z-index: 3;
      }
    }

    &:first-child > .btn + .btn {
      margin-left: 0;
    }

    /* Negative margin to only have a single, shared border between the two */
    &:not(:last-child) {
      > .btn,
      > .btn-group {
        margin-right: -1px;
      }
    }
    &:not(:first-child) {
      > .btn,
      > .btn-group {
        z-index: 2;
        /* remove nagative margin to solve overlapping issue with button. */
        margin-left: 0;

        /* When input is first, overlap the right side of it with the button(-group) */
        &:first-child {
          margin-left: -1px;
        }

        /* Because specificity */
        &:focus,
        &:active,
        &:hover {
          z-index: 3;
        }
      }
    }

    &.dropdown {
      display: table-cell;
    }

    .btn {
      white-space: nowrap;
    }
  }
}
`;
