// @ux/style-common/src/media.js
import { css } from 'linaria';
import { mediaBreakpointUp } from './mixins/breakpoints';
import intents from './intents.config.js';

export default css`
/* To be updated when grid gets updated. */
:global() {
  /* Images/elements as the media anchor */

  .media-object {
    display: block;

    /* Fix collapse in webkit from max-width: 100% and display: table-cell. */
    &.img-thumbnail {
      max-width: none;
    }
  }



  /* Alignment */

  .media-right {
    padding-left: calc(${intents.ux.box.density} * 2.5);
  }

  .media-left {
    padding-right: calc(${intents.ux.box.density} * 2.5);
  }


  /* Headings */

  .media-heading {
    margin-top: 0;
    margin-bottom: calc(${intents.ux.box.density} * 2);
  }


  /* Media list variation */

  .media-list {
    padding-left: 0;
    list-style: none;
  }

  /* Media Small*/

  .media-small {
    .media-body {
      vertical-align: middle;
    }
    .media-heading {
      margin: 0;
    }
    .media-heading + p {
      margin: calc(${intents.ux.box.density} * 2) 0 0;
    }
  }


  /* Play buttons */

  .play-button {
    border-radius: 50%;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    position: relative;
    &:before {
      content: '';
      width: 0;
      height: 0;
      border-top: .3125rem solid transparent;
      border-left: .5rem solid currentColor;
      border-bottom: .3125rem solid transparent;
      left: 50%;
      margin-right: -.125rem;
      z-index: 1;
    }
    &:after {
      content: '';
      position: absolute;
      width: 0;
      height: 0;
      border-top: .1875rem solid transparent;
      border-left: .3125rem solid currentColor;
      border-bottom: .1875rem solid transparent;
      filter: invert(1);
      left: 50%;
      margin-left: -.125rem;
      z-index: 2;
    }
    &.inline {
      width: 1.5rem;
      height: 1.5rem;
      margin-right: .5rem;
      ${mediaBreakpointUp('lg', `
        width: 2rem;
        height: 2rem;
      `)}
    }
    &.video {
      width: 3.5rem;
      height: 3.5rem;
      margin: .5rem;
      &:before {
        border-top-width: .5rem;
        border-left-width: .75rem;
        border-bottom-width: .5rem;
        margin-right: -.3125rem;
      }
      &:after {
        border-top-width: .375rem;
        border-left-width: .5625rem;
        border-bottom-width: .375rem;
      }
      ${mediaBreakpointUp('lg', `
        width: 88px;
        height: 88px;
        &:before {
          border-top-width: .5rem;
          border-left-width: .75rem;
          border-bottom-width: .5rem;
          margin-right: -.375rem;
        }
        &:after {
          border-top-width: .375rem;
          border-left-width: .5625rem;
          border-bottom-width: .375rem;
        }
      `)}
    }
    &:not([class*='bg-']) {
      &:before {
        color: ${intents.ux.action.foregroundColor};
      }
      &.inline {
        background-color: transparent;
        border: 1px solid currentColor;
      }
      &.video {
        background-color: ${intents.ux.box.backgroundColor};
      }
    }
  }
}
`;
