// @ux/style-common/src/nav.js
import { css } from 'linaria';
import intents from './intents.config.js';

export default css`
:global() {
  /* Base class */

  /* Kickstart any navigation component with a set of style resets. Works with
  '<nav>'s or '<ul>'s. */

  .nav {
    font-family: ${intents.ux.text.fontFamily};
    font-weight: ${intents.ux.text.fontWeight};
    font-size: calc(${intents.ux.text.fontSize} * .875);
    padding-left: 0;
    margin-bottom: 0;
    list-style: none;
    display: flex;
    flex-wrap: nowrap;
  }

  .nav-item {
    margin: 0;
  }

  .nav-link {
    color: ${intents.ux.box.foregroundColor};
    display: inline-block;
    padding: calc(${intents.ux.box.density} * 1) calc(${intents.ux.box.density} * 8);
    text-decoration: none;

    &:focus,
    &:hover {
      color: ${intents.ux.boxLowContrast.foregroundColor};
      background-color: ${intents.ux.boxLowContrast.backgroundColor};
      text-decoration: none;
      outline: none;
    }

    /* Disabled state lightens text and removes hover/tab effects */
    &.disabled {
      opacity: .4;
      &,
      &:focus,
      &:hover {
        cursor: not-allowed;
        background-color: transparent;
      }
    }
  }


  /* Nav inline */

  .nav-inline {
    .nav-item {
      display: inline-block;
    }

    .nav-item + .nav-item,
    .nav-link + .nav-link {
      margin-left: calc(${intents.ux.box.density} * 4);
    }
  }


  /* Pills */

  .nav-pills {
    &::after {
      content: '';
      display: table;
      clear: both;
    }

    .nav-item {
      float: left;

      + .nav-item {
        margin-left: .2rem;
      }
    }

    .nav-link {
      display: block;
      padding: calc(${intents.ux.box.density} * 1) calc(${intents.ux.box.density} * 8);
      border-radius: ${intents.ux.action.borderRadius};
    }

    .nav-link.active,
    .nav-item.open .nav-link {
      &,
      &:focus,
      &:hover {
        color: ${intents.ux.feedbackHighlight.onFeedbackColor};
        background-color: ${intents.ux.feedbackHighlight.feedbackColor};
        cursor: default;
      }
    }
  }

  .nav-stacked {
    .nav-item {
      display: block;
      float: none;

      + .nav-item {
        margin-top: .2rem;
        margin-left: 0;
      }
    }
  }


  /* Tabbable tabs */


  /* Hide tabbable panes to start, show them when '.active' */
  .tab-content {
    > .tab-pane {
      display: none;
    }
    > .active {
      display: block;
    }
  }
}
`;
