// @ux/style-common/src/type.js
import { css } from 'linaria';
import intents from './intents.config.js';
import { mediaBreakpointUp } from './mixins/breakpoints';
import { ms } from './constants';


export default css`
:global() {
  *::selection {
    background: ${intents.ux.feedbackHighlight.feedbackColor}; /* palette('primary', 'highlight' */
    color: ${intents.ux.feedbackHighlight.onFeedbackColor}; /* palette('primary', 'highlight', 'contrast'; */
  }

  .font-base {
    font-family: ${intents.ux.text.fontFamily};
  }

  .font-primary-bold {
    font-family: ${intents.ux.text.fontFamily};
    font-weight: bolder;
  }

  /* Headings */

  /* stylelint-disable selector-list-comma-newline-after */
  h1, h2, h3, h4, h5, h6,
  .h1, .h2, .h3, .h4, .h5, .h6 {
    margin-bottom: calc(${intents.ux.box.density} * 2);
    font-family: ${intents.ux.textTitle.fontFamily};
    font-weight: ${intents.ux.textTitle.fontWeight};
    color: currentColor;

    &.margin-none {
      margin-bottom: 0;
    }
    &.margin-compressed {
      margin-bottom: calc(${intents.ux.box.density} * 2);
    }
  }

  .h0 {
    font-family: ${intents.ux.textHeading.fontFamily};
    font-weight: bolder !important;
    font-size: ${ms(6)}; /* 2.027rem */
    line-height: ${ms(8)}; /* 2.566rem */
    margin-bottom: calc(${intents.ux.box.density} * 4);

    ${mediaBreakpointUp('lg', `
      font-size: ${ms(8)}; /* 2.566rem */
      line-height: ${ms(10)}; /* 3.247rem */
    `)}
  }

  h1, .h1 {
    font-weight: ${intents.ux.textTitle.fontWeight};
    font-size: ${ms(6)}; /* 2.027rem */
    line-height: 2.25rem;
    margin-bottom: calc(${intents.ux.box.density} * 4);

    ${mediaBreakpointUp('lg', `
      font-size: ${ms(7)}; /* 2.281rem */
      line-height: 2.5rem;
    `)}

    /* Brand heading 1 */
    &.headline-brand {
      font-size: ${ms(7)}; /* 2.281rem */
      line-height: 2.75rem;
      ${mediaBreakpointUp('lg', `
        font-size: ${ms(9)}; /* 2.887rem */
        line-height: 3.5rem;
      `)}
      ${mediaBreakpointUp('xxl', `
        font-size: ${ms(10)}; /* 3.247rem */
        line-height: 4rem;
      `)}
    }
  }

  h2, .h2 {
    font-weight: bolder;
    font-size: ${ms(5)}; /* 1.802rem */
    line-height: 2.25rem;
    margin-bottom: calc(${intents.ux.box.density} * 4);

    ${mediaBreakpointUp('lg', `
      font-size: ${ms(6)}; /* 2.027rem */
      line-height: 2.5rem;
    `)}

    /* Brand heading 2 */
    &.headline-brand {
      font-size: ${ms(6)};
      line-height: 2.5rem;
      ${mediaBreakpointUp('lg', `
        font-size: ${ms(8)}; /* 2.566rem */
        line-height: 3.25rem;
      `)}
      ${mediaBreakpointUp('xxl', `
        font-size: ${ms(9)}; /* 2.887rem */
        line-height: 3.5rem;
      `)}
    }
  }

  h3, .h3 {
    font-size: ${ms(4)}; /* 1.602rem */
    line-height: 2rem;

    ${mediaBreakpointUp('lg', `
      font-size: ${ms(5)}; /* 1.802rem */
      line-height: 2.25rem;
    `)}

    /* Brand heading 3 */
    &.headline-brand {
      font-size: ${ms(5)}; /* 1.802rem */
      line-height: 2.25rem;
      ${mediaBreakpointUp('lg', `
        font-size: ${ms(7)}; /* 2.281rem */
        line-height: 2.75rem;
      `)}
      ${mediaBreakpointUp('xxl', `
        font-size: ${ms(8)}; /* 2.566rem */
        line-height: 3.25rem;
      `)}
    }
  }

  h4, .h4 {
    font-size: ${ms(3)}; /* 1.424rem */
    line-height: 1.75rem;

    ${mediaBreakpointUp('lg', `
      font-size: ${ms(4)}; /* 1.602rem */
      line-height: 2rem;
    `)}

    /* Brand heading 4 */
    &.headline-brand {
      font-size: ${ms(3)}; /* 1.424rem */
      line-height: 2rem;

      ${mediaBreakpointUp('lg', `
        font-size: ${ms(6)}; /* 2.027rem */
        line-height: 2.5rem;
      `)}
      ${mediaBreakpointUp('xxl', `
        font-size: ${ms(7)}; /* 2.281rem */
        line-height: 2.75rem;
      `)}
    }
  }

  h5, .h5 {
    font-size: ${ms(2)}; /* 1.266rem */
    line-height: 1.75rem;
  }

  h6, .h6 {
    font-size: ${ms(1)}; /* 1.125rem */
    line-height: 1.5rem;
  }

  [class*="bg-"] {
    h1, h2, h3, h4, h5, h6,
    .h0, .h1, .h2, .h3, .h4, .h5, .h6 {
      color: inherit !important;
    }
  }

  .display-1,
  .display-2,
  .display-3,
  .display-4,
  .headline-brand {
    font-family: ${intents.ux.textHeading.fontFamily};
    font-weight: bolder !important;
  }

  /* Display classes - intents come back to */

  .display-1 {
    font-size: ${ms(10)}; /* 3.247rem */
    line-height: 3.5rem; /* 3.5rem */
    font-weight: bolder;
    ${mediaBreakpointUp('lg', `
      font-size: ${ms(13)}; /* 4.624rem */
      line-height: 5.25rem;
    `)}
    ${mediaBreakpointUp('xxl', `{
      font-size: ${ms(16)};
      line-height: 7.25rem;
    }`)}
  }

  .display-2 {
    font-size: ${ms(8)}; /* 2.566rem */
    line-height: 3rem;
    font-weight: bolder;
    ${mediaBreakpointUp('lg', `
      font-size: ${ms(12)}; /* 4.11rem */
      line-height: 4.75rem;
    `)}
    ${mediaBreakpointUp('xxl', `
      font-size: ${ms(15)}; /* 5.852rem */
      line-height: 6.5rem;
    `)}
  }

  .display-3 {
    font-size: ${ms(8)}; /* 2.566rem */
    line-height: 3rem;
    font-weight: bolder;
    ${mediaBreakpointUp('lg', `
      font-size: ${ms(12)}; /* 4.11rem */
      line-height: 4.375rem;
    `)}
    ${mediaBreakpointUp('xxl', `
      font-size: ${ms(15)}; /* 5.852rem */
      line-height: 5.25rem;
    `)}
  }

  .display-4 {
    font-size: ${ms(8)}; /* 2.566rem */
    line-height: 3rem;
    font-weight: bolder;
    ${mediaBreakpointUp('lg', `
      font-size: ${ms(10)}; /* 3.247rem */
      line-height: 4.75rem;
    `)}
    ${mediaBreakpointUp('xxl', `
      font-size: ${ms(12)}; /* 4.11rem */
    `)}
  }
  /* stylelint-enable selector-list-comma-newline-after */



  /* Body text */

  .lead {
    font-size: ${intents.ux.textTitle.fontSize};
    font-weight: ${intents.ux.text.fontWeight};
  }

  .brand-message {
    font-family: ${intents.ux.text.fontFamily};
    font-size: ${intents.ux.textTitle.fontSize};
    font-weight: bolder;
  }


  /* Horizontal rules */

  hr {
    border: 0;
    border-top: 1px solid ${intents.ux.boxLowContrast.borderColor};
    margin-top: calc(${intents.ux.box.density} * 4);
    margin-bottom: calc(${intents.ux.box.density} * 4);
  }

  .ux-hr-dark {
    border-top: 1px solid ${intents.ux.feedbackHighContrastPassive.feedbackColor};
  }


  /* Lists */

  .list-unstyled {
    padding-left: 0;
    list-style: none;
  }

  /* Inline turns list items into inline-block */
  .list-inline {
    padding-left: 0;
    list-style: none;
  }
  .list-inline-item {
    display: inline-block;

    &:not(:last-child) {
      margin-right: calc(${intents.ux.box.density} * 1);
    }
  }

  /* Misc */

  small,
  .small {
    font-size: ${intents.ux.text.fontSize};
    font-weight: ${intents.ux.text.fontWeight};
  }

  .large {
    font-size: ${ms(1)}; /* 1.125rem */
  }

  .legal {
    font-size: ${ms(-2)}; /* .79rem */
  }

  mark,
  .mark {
    display: inline-block;
    padding: .08em .10em;
    margin: 0;
    line-height: ${intents.ux.text.lineHeight};
    background: ${intents.ux.feedbackHighlight.feedbackColor});
  }


  /* Builds on 'abbr' */
  .initialism {
    font-size: 90%;
    text-transform: uppercase;
  }

  /* Blockquotes */
  .blockquote {
    border-left: 4px solid ${intents.ux.boxLowContrast.borderColor};
    padding: calc(${intents.ux.box.density} * 2) calc(${intents.ux.box.density} * 4);
    margin-bottom: calc(${intents.ux.box.density} * 4);
    font-size: calc(${intents.ux.text.fontSize} * 1.25);
  }

  .blockquote-footer {
    color: ${intents.ux.box.foregroundColor};
    display: block;
    font-size: 80%; /* back to default font-size */

    &::before {
      content: "\\2014 \\00A0"; /* em dash, nbsp */
    }
  }

  /* Opposite alignment of blockquote */
  .blockquote-reverse {
    border-right: 4px solid ${intents.ux.boxLowContrast.borderColor};
    padding-right: calc(${intents.ux.box.density} * 4);
    padding-left: 0;
    text-align: right;
    border-left: 0;
  }

  .blockquote-reverse .blockquote-footer {
    &::before {
      content: "";
    }
    &::after {
      content: "\\00a0 \\2014"; /* nbsp, em dash */
    }
  }

  /* Link styles */
  /* solo, action links */
  .standalone-link {
    font-size: ${intents.ux.text.fontSize};
    font-weight: bolder;
    position: relative;
    text-decoration: none;
    padding-right: 1.75rem !important;
    &.action {
      &:before {
        content: '';
        background-color: ${intents.ux.feedbackHighlight.feedbackColor};
        height: 2px;
        width: 1.1rem;
        display: inline-block;
        position: absolute;
        right: .125rem;
        top: 50%;
        margin-top: -1px;
      }
      &:after {
        content: '';
        border-color: inherit;
        border-width: 0 2px 2px 0;
        border-style: solid;
        display: inline-block;
        transform: rotate(-45deg);
        padding: .25rem;
        position: absolute;
        right: .125rem;
        top: 50%;
        margin-top: -5px;
      }
      &:hover {
        &:after {
          text-decoration: none;
        }
        &:before {
          background-color: ${intents.ux.feedbackHighlight.feedbackColor};
        }
      }
    }
    &.lg {
      font-size: ${ms(1)}; /* 1.125rem */
    }
    &.sm {
      font-size: ${ms(-1)};
    }
  }

  /* Paragraph styles */

  .p-lg {
    font-size: ${ms(1)};
    line-height: 1.75rem; /* 28px */
  }

  .p-base {
    font-size: ${intents.ux.textParagraph.fontSize};
    line-height: ${intents.ux.textParagraph.lineHeight};
  }

  .p-sm {
    font-size: ${ms(-1)};
    line-height: 1.25rem; /* 20px */
  }

  .p-xs {
    font-size: ${ms(-2)};
    line-height: 1.25rem; /* 20px */
  }


  /* Titles */

  .title { /* Base size 16px, line-height 24px etc */
    font-size: ${intents.ux.textTitle.fontSize};
    line-height: ${intents.ux.textTitle.lineHeight};
    color: ${intents.ux.box.foregroundColor};
    margin-bottom: calc(${intents.ux.box.density} * 3);
    font-weight: bolder;
    &.extra-large {
      font-size: ${ms(2)}; /* 1.266rem */
      line-height: 1.75rem;
    }
    &.large {
      font-size: ${ms(1)};
    }
    &.small {
      line-height: ${intents.ux.textTitle.lineHeight};
      color: ${intents.ux.boxLowContrast.foregroundColor};
      font-size: ${ms(-2)};
      text-transform: uppercase;
      letter-spacing: .05625rem;
    }
    &.extra-small {
      font-size: ${ms(-4)};
      line-height: 1rem;
      text-transform: uppercase;
      letter-spacing: .05625rem;
    }
  }
}
`;
