// @ux/progress-steps/lib/progress-step.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ProgressStep;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _checkmark = _interopRequireDefault(require("@ux/icon/checkmark"));

/**
 * Possible status types
 *
 * @type {object}
 */
const STATUS = {
  Incomplete: 'incomplete',
  Current: 'current',
  Complete: 'complete'
};
/**
 * ProgressStep component, representing each individual step.
 *
 * @param {Object} props Props
 * @returns {ReactElement} ReactElement
 */

function ProgressStep(props) {
  return /*#__PURE__*/_react.default.createElement("li", {
    role: "status",
    "aria-live": "polite",
    "aria-label": `step ${props.status}`,
    className: `ux-tabs-wiz-step-${props.status}`
  }, /*#__PURE__*/_react.default.createElement("a", {
    href: props.href,
    onClick: props.onClick,
    className: 'ux-tabs-wiz-link',
    tabIndex: props.href ? null : props.tabIndex
  }, props.status === 'complete' && /*#__PURE__*/_react.default.createElement(_checkmark.default, {
    fill: "white"
  }), /*#__PURE__*/_react.default.createElement("span", {
    className: "ux-tabs-wiz-txt"
  }, props.children)));
}

ProgressStep.propTypes = {
  status: _propTypes.default.string.isRequired,
  href: _propTypes.default.string,
  onClick: _propTypes.default.func,
  children: _propTypes.default.node,
  tabIndex: _propTypes.default.oneOf([-1, 0])
};
ProgressStep.defaultProps = {
  status: STATUS.Incomplete,
  tabIndex: 0
};
ProgressStep.Status = STATUS;
