// @ux/collapsible/src/styles.js

import { css } from 'linaria';
import intents from './intents.config.js';
import '@ux/text/src/styles';

export default css`
:global() {
  @import '~@ux/icon/chevron-down/index.css';

  .panel {
    border-bottom: 1px solid ${intents.uxCollapsible.borderColor};
    background-color: ${intents.uxCollapsible.backgroundColor};
    color: ${intents.uxCollapsible.foregroundColor};
    display: flex;
    align-items: stretch;
    flex-direction: column;
  }

  .panel-button {
    border: 0;
    background-color: transparent;
    text-decoration: none;
    display: flex;
    align-items: center;
    cursor: pointer;
    position: relative;
    outline: none;

    .panel-title {
      margin-bottom: 0;
      font-family: ${intents.uxCollapsible.title.fontFamily};
      font-weight: ${intents.uxCollapsible.title.fontWeight};
      line-height: ${intents.uxCollapsible.title.lineHeight};
      color: ${intents.uxCollapsible.title.foregroundColor};
      padding: calc(${intents.ux.box.density} * 8) 0;
      text-align: left; /* coutneracts button centering */
    }

    .panel-icon {
      display: inline-flex;
      margin-left: auto;
      transform: rotate(180deg);
      transform-origin: center center;
      transition: transform 300ms ease;
      color: ${intents.uxCollapsible.title.foregroundColor};
    }

    &[aria-expanded=false],
    &.collapsed .panel-icon {
      transform: rotate(0deg);
    }

    .panel-indicator {
      width: 1rem;
      align-self: stretch;
      margin-right: calc(${intents.ux.box.density} * 4);
      display: flex;
      flex-grow: 0;
      align-items: center;
      justify-content: center;
    }
  }

  .panel-collapse {
    margin: 0;
    color: ${intents.uxCollapsible.content.foregroundColor};
    background-color: ${intents.uxCollapsible.content.backgroundColor};
    opacity: 1;
    overflow-y: hidden; /* needed for transition */
    transition: height .35s ease;

    &.collapse.in {
      overflow-y: auto;
    }
  }

  .panel-content {
    margin: calc(${intents.ux.box.density} * 8);

    > *:last-child {
      margin-bottom: 0;
    }
  }
}
`;
