// @ux/table/src/styles.js
import { css } from 'linaria';
import { mixins, variables } from '@ux/stylevars';
import intents from './intents.config.js';

const {
  mediaBreakpointDown,
  mediaBreakpointUp,
  breakpointNext,
  breakpointInfix
} = mixins;
const {
  gridBreakpoints
} = variables;

function tableResponsive(breakpoints, spacer) {
  return Object.keys(breakpoints).map((breakpoint) => {
    const next = breakpointNext(breakpoint, breakpoints);
    const infix = breakpointInfix(next, breakpoints);
    if (next) {
      return `
        &${infix} {
          ${mediaBreakpointDown(breakpoint, `
            display: block;
            width: 100%;
            overflow-x: auto;
            margin-bottom: ${spacer};
            &.table-bordered {
              border: 0;
            }
        `)}
        }
      `;
    }
    return `
      ${mediaBreakpointDown(breakpoint, `
        display: block;
        width: 100%;
        overflow-x: auto;
        margin-bottom: ${spacer};
        &.table-bordered {
          border: 0;
        }
    `)}
    `;
  }).join('');
}

export default css`
:global() {
  @import '~@ux/icon/chevron-down/index.css';
  @import '~@ux/icon/chevron-up/index.css';
  /* stylelint-disable max-nesting-depth */

  .ux-table,
  .basegrid {
    width: 100%;
    max-width: 100%;
    margin-bottom: calc(${intents.ux.box.density} * 4);
    word-wrap: break-word;
    border-collapse: collapse;

    tr {
      background-color: ${intents.uxTable.item.backgroundColor};
    }

    tr:last-child {
      border-bottom: 1px solid ${intents.uxTable.item.borderColor};
    }

    th,
    td {
      padding: calc(${intents.ux.box.density} * 4);
      vertical-align: top;
      font-size: ${intents.uxTable.item.fontSize};
      line-height: ${intents.uxTable.item.lineHeight};
      font-family: ${intents.uxTable.item.fontFamily};
      font-weight: ${intents.uxTable.item.fontWeight};
      color: ${intents.uxTable.item.foregroundColor};
      border-top: 1px solid ${intents.uxTable.item.borderColor};
    }

    /* Bottom align for column headings */
    > thead > tr > th {
      font-size: ${intents.uxTable.label.fontSize};
      line-height: ${intents.uxTable.label.lineHeight};
      font-family: ${intents.uxTable.label.fontFamily};
      font-weight: ${intents.uxTable.label.fontWeight};
      vertical-align: middle;
      color: ${intents.uxTable.label.foregroundColor};
      background-color: ${intents.uxTable.label.backgroundColor};
    }
    /* Remove top border from thead by default */
    > caption + thead,
    > colgroup + thead,
    > thead:first-child {
      > tr:first-child {
        > th,
        > td {
          border-top: 0 !important;
        }
      }
    }

    /* Account for multiple tbody instances */
    > tbody + tbody {
      border-top: 2px solid ${intents.uxTable.item.borderColor};
    }

    svg {
      width: 1.375rem;
    }
  }

  /* Expanded table w/ double padding */
  .ux-table.table-expanded {
    > thead,
    > tbody,
    > tfoot {
      > tr {
        > th,
        > td {
          padding-top: calc(${intents.ux.box.density} * 8);
          padding-bottom: calc(${intents.ux.box.density} * 8);
        }
      }
    }
  }


  /* Condensed table w/ half padding */

  .ux-table.table-sm {
    th,
    td {
      padding: calc(${intents.ux.box.density} * 1);
    }
  }


  /* Zebra-striping */
  /* Default zebra-stripe styles (alternating gray and transparent backgrounds) */

  .ux-table.table-striped,
  .basegrid {
    tbody tr:nth-of-type(odd) {
      background-color: ${intents.uxTable.item.stripeColor};
    }
  }

  /* Responsive tables */
  /* Generate series of '.table-responsive-*' classes for configuring the screen */
  /* size of where your table will overflow. */
  /* Do we really need these going to uxcore(3) next? */

  .ux-table.table-responsive {
    ${tableResponsive(gridBreakpoints, '1rem')}
    > .table {
      margin-bottom: 0;
    }
    ${mediaBreakpointDown('md', `
      width: 100%;
      margin-bottom: calc(${intents.ux.box.density} * 6);
      overflow-y: hidden;
      border: 1px solid ${intents.uxTable.borderColor};
      /* Special overrides for the bordered tables */
      > .table { /* GD! */
        border: 0;

        /* Nuke the appropriate borders so that the parent can handle them */
        > thead,
        > tbody,
        > tfoot {
          > tr {
            > th:first-child,
            > td:first-child {
              border-left: 0;
            }
            > th:last-child,
            > td:last-child {
              border-right: 0;
            }
          }
        }

        /* Only nuke the last row's bottom-border in 'tbody' and 'tfoot' since */
        /* chances are there will be only one 'tr' in a 'thead' and that would */
        /* remove the border altogether. */
        > tbody,
        > tfoot {
          > tr:last-child {
            > th,
            > td {
              border-bottom: 0;
            }
          }
        }
      }
    `)}
  }

  /* GD Basegrid */
  /* Makes a striped, hovered, and bordered table */
  /* without all the classes */
  .basegrid {
    table-layout: fixed;

    > thead > tr > th {
      height: 5.714rem;
      ${mediaBreakpointUp('md', `
        height: 5rem;
      `)}
    }
  }

  /* GD Responsive Tables */
  .ux-table.table-responsive {
    .basegrid {
      ${mediaBreakpointDown('md', `
        max-width: none;
        border: 0;
      `)}
      ${mediaBreakpointUp('lg', `width: 100% !important; /* to override mobile in page style */`)}
    }
  }

  /* GD Features */
  /* Sets all content centered horizontally and vertically */
  .ux-table.table-features {
    > thead > tr,
    > tbody > tr {
      > th,
      > td {
        text-align: center;
        vertical-align: middle;
      }
      .table-feature {
        text-align: left;
      }
    }
  }


  ${mediaBreakpointDown('md', `
    /* makes td/th stack on top of each other */
    .ux-table.table-stacking {
      &,
      > tbody,
      > tbody > tr,
      > tbody > tr > td {
        display: block !important;
        width: 100%;
      }
      > thead {
        display: none;
      }
      > tbody > tr:first-child > td:first-child {
        border-top: 0;
      }
      > tbody > tr > td {
        padding-left: 10rem;
        text-align: left;
        vertical-align: top;
        border-right: 0;
        border-bottom: 0;
        border-left: 0;

        &:before {
          display: block;
          float: left;
          width: 10rem;
          padding-right: calc(${intents.ux.box.density} * 4);
          padding-left: calc(${intents.ux.box.density} * 4);
          margin-left: -10rem;
          content: attr(data-th);
          font-family: ${intents.uxTable.label.fontFamily};
          font-weight: ${intents.uxTable.label.fontWeight};
        }
      }
    }
  `)}

  .ux-table-sort {
    padding-right: 15px;
    &,
    &:hover,
    &:focus {
      color: inherit;
      text-decoration: none;
    }

    &:hover {
      cursor: pointer;
      .drop-up-down-icon {
        visibility: visible !important;
      }
    }

    .drop-up-down-icon svg {
      position: relative;
      margin-top: -1px;
      margin-left: 0.5em;
      vertical-align: top;
    }
  }
}
`;
