// @ux/pivot/examples/default.stories.js
import Pivots from '../src';
import '../src/styles';
import React from 'react';
import Settings from '@ux/icon/settings';
import '@ux/icon/settings/styles';

export default function DefaultExample() {
  const data = [
    {
      title: 'My title',
      href: '#',
      graphic: <img alt='placeholder' src='http://placehold.it/50x50' />,
      subtitle: 'This is an image tag and here is some longer text to test wrapping',
      id: 'example-1'
    },
    {
      title: 'My title',
      href: '#',
      graphic: <Settings />,
      subtitle: 'This is a font icon',
      id: 'example-2'
    },
    {
      title: 'Arrow Tile',
      href: '#',
      arrow: true,
      subtitle: 'This is an arrow',
      id: 'example-3'
    },
    {
      title: 'My title',
      href: '#',
      graphic:
        <svg version='1.1' xmlns='http://www.w3.org/2000/svg' xmlnsXlink='http://www.w3.org/1999/xlink'
          width='32' height='32' viewBox='0 0 32 32'>
          <path fill='rgb(119, 192, 67)'
            d='M28 16c0 6.627-5.373 12-12 12s-12-5.373-12-12c0-6.627 5.373-12 12-12s12 5.373 12 12z'></path>
          <path fill='rgb(51, 51, 51)' d='M25 16c0 4.971-4.029 9-9 9s-9-4.029-9-9c0-4.971 4.029-9 9-9s9 4.029 9 9z'></path>
          <path fill='rgb(119, 192, 67)' d='M0 13h6v6h-6v-6z'></path>
          <path fill='rgb(119, 192, 67)' d='M26 13h6v6h-6v-6z'></path>
          <path fill='rgb(119, 192, 67)' d='M13 26h6v6h-6v-6z'></path>
          <path fill='rgb(119, 192, 67)' d='M13 0h6v6h-6v-6z'></path>
          <path fill='rgb(119, 192, 67)' d='M25.295 1.042l4.738 4.738-4.243 4.243-4.738-4.738 4.243-4.243z'></path>
          <path fill='rgb(119, 192, 67)' d='M10.973 5.295l-4.738 4.738-4.243-4.243 4.738-4.738 4.243 4.243z'></path>
          <path fill='rgb(119, 192, 67)' d='M10.962 26.27l-4.738 4.738-4.243-4.243 4.738-4.738 4.243 4.243z'></path>
          <path fill='rgb(119, 192, 67)' d='M25.285 22.038l4.738 4.738-4.243 4.243-4.738-4.738 4.243-4.243z'></path>
          <path fill='rgb(255, 255, 255)' d='M22 16c0 3.314-2.686 6-6 6s-6-2.686-6-6c0-3.314 2.686-6 6-6s6 2.686 6 6z'></path>
        </svg>,
      subtitle: 'This is an inline svg',
      id: 'example-4'
    }
  ];

  return (
    <Pivots pivotList={ data } grid={{ md: 6, lg: 4 }} />
  );
}
