// @ux/pivot/examples/toggle-tiles.stories.js
import Pivots from '../src';
import '../src/styles';
import React, { useState } from 'react';

export default function ToggleTiles() {
  const [selectedIndex, setIndex] = useState(1);

  function getData(group) {
    return ['1st', '2nd', '3rd'].map(name => ({
      title: `${name} for ${group}`,
      onClick: () => console.log(`clicked ${name} as button, not an anchor`),
      graphic: <img alt='placeholder' src='http://placehold.it/50x50' />,
      subtitle: 'This is an image tag'
    }));
  }

  return <div className='card ux-card'>
    <div className='card-block'>
      <p>First Grouping</p>
      <Pivots
        pivotList={ getData('Group One') }
        selectedIndex={ selectedIndex }
        grid={{ md: 6, lg: 4 }}
      />
      <p>
        <br/>
        <button
          className='btn btn-primary'
          onClick={ () => setIndex(-1) }>
          Reset first group selection
        </button>
        <br/>
      </p>
      <p><br/>Second Grouping</p>
      <Pivots
        pivotList={ getData('Group Two') }
        grid={{ md: 6, lg: 4 }}
      />
    </div>
  </div>;
}
