// @ux/style-common/webpack.config.js
const path = require('path');
const MiniCssExtractPlugin = require('mini-css-extract-plugin');
const config = require('@ux/webpack-config');

module.exports = {
  name: 'linaria',
  mode: 'production',
  target: 'browserslist',
  entry: path.join(__dirname, 'src', 'index.js'),
  output: {
    path: '/dev',
    filename: 'null'
  },
  plugins: [
    new MiniCssExtractPlugin({
      filename: path.join('..', __dirname, 'lib', 'style-common.css')
    })
  ],
  module: {
    rules: [{
      test: /\.jsx?$/,
      exclude: /external-helpers\.js/,
      use: [{
        loader: 'babel-loader'
      }, {
        loader: 'linaria/loader',
        options: { sourceMap: false }
      }]
    }, {
      test: /\.s?css$/,
      use: [
        {
          loader: MiniCssExtractPlugin.loader
        },
        {
          loader: 'css-loader',
          options: {
            importLoaders: 1
          }
        },
        {
          loader: 'postcss-loader',
          options: {
            postcssOptions: {
              plugins: config.postcss
            }
          }
        }
      ]
    }]
  }
};
