// @ux/stylevars/src/functions/button.js
import { math } from 'polished';
import variables from '../variables';

function buttonGet(variant, property) {
  return variables.buttons[variant][property];
}

function buttonPadding(paddingY, paddingX) {
  return `${paddingY} ${paddingX}`;
}

function buttonPaddingSm(paddingY, paddingX) {
  return `${math(`${paddingY} + 1`)} ${paddingX} ${math(`${paddingY} - 1`)}`;
}

function buttonPaddingDefaults(paddingY, paddingX, btnBorderDefaults) {
  return `${math(`${paddingY} - ${btnBorderDefaults}`)} ${math(`${paddingX} - ${btnBorderDefaults}`)}`;
}

function buttonPaddingDefaultsSM(paddingY, paddingX, btnBorderDefaults) {
  // eslint-disable-next-line max-len
  return `${math(`${paddingY} - ${btnBorderDefaults} + 1`)} ${math(`${paddingX} - ${btnBorderDefaults}`)} ${math(`${paddingY} - ${btnBorderDefaults} - 1`)}`;
}

function buttonPaddingCanvas(paddingY, paddingX) {
  return `${paddingY} ${paddingX}`;
}

export {
  buttonGet,
  buttonPadding,
  buttonPaddingSm,
  buttonPaddingDefaults,
  buttonPaddingCanvas,
  buttonPaddingDefaultsSM
};
