// @ux/text-entry/src/styles.js
import { css } from 'linaria';
import intents from './intents.config.js';

import '@ux/text/src/styles';
import '@ux/label/src/styles';

export default css`
  :global() {
    .ux-text-entry-shell {
      --uxTextEntry-padding: calc(${intents.ux.box.density} * 2);
      position: relative;
      flex: 1;
      white-space: nowrap;
      text-overflow: ellipsis;
      overflow: hidden;
      padding: var(--uxTextEntry-padding);
      border: var(--uxButton--borderWidth, 2px) solid transparent; /* Match with standard button border */

      &[data-size]:after {
        /** When a label isn't present, we need to set the height of the component with this character */
        content: attr(data-size);
        visibility: hidden;
        white-space: pre;
      }

      .ux-label ~ .ux-text-entry-field {
        transform: translateY(var(--uxTextEntry-padding));
      }
    }

    .ux-text-entry-field,
    .ux-text-entry-suggestion {
      display: flex;
      align-items: center;
      position: absolute;
      height: calc(100% - (var(--uxTextEntry-padding) * 2));
      width: calc(100% - (var(--uxTextEntry-padding) * 2));
      top: 0;
      left: 0;
      font: inherit;
      color: inherit;
      text-align: revert;
      background-color: transparent;
      margin: var(--uxTextEntry-padding);
      padding: 0;
      border: 0;
    }

    .ux-text-entry-field {
      resize: none;

      &::placeholder {
        color: ${intents.uxTextEntry.placeholder.foregroundColor};
        opacity: 1;
      }

      /* Hide webkit accessories */
      &::-webkit-search-cancel-button,
      &::-webkit-inner-spin-button {
        display: none;
      }

      /* Hide moz accessories */
      &[type=number] {
        -moz-appearance: textfield;
      }

      /* Hide ms accessories */
      &::-ms-reveal,
      &::-ms-clear {
        display: none;
      }

      &:focus {
        outline: 0;
      }

      &[disabled] {
        opacity: .4; /* disabled opacity */
        cursor: not-allowed;
      }

      &[read-only] {
        cursor: not-allowed;
      }
    }

    .ux-text-entry-suggestion {
      white-space: nowrap;
      text-overflow: clip;
      overflow: hidden;
      opacity: .4;
    }
  }
`;
