// @ux/uxcore2/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Accordion", {
  enumerable: true,
  get: function () {
    return _accordion.default;
  }
});
Object.defineProperty(exports, "Alert", {
  enumerable: true,
  get: function () {
    return _alert.default;
  }
});
Object.defineProperty(exports, "Button", {
  enumerable: true,
  get: function () {
    return _button.default;
  }
});
Object.defineProperty(exports, "Checkbox", {
  enumerable: true,
  get: function () {
    return _checkbox.default;
  }
});
Object.defineProperty(exports, "Collapsible", {
  enumerable: true,
  get: function () {
    return _collapsible.default;
  }
});
Object.defineProperty(exports, "Criteria", {
  enumerable: true,
  get: function () {
    return _criteria.default;
  }
});
Object.defineProperty(exports, "Dropdown", {
  enumerable: true,
  get: function () {
    return _dropdown.default;
  }
});
Object.defineProperty(exports, "Fieldset", {
  enumerable: true,
  get: function () {
    return _fieldset.default;
  }
});
Object.defineProperty(exports, "Growl", {
  enumerable: true,
  get: function () {
    return _growl.default;
  }
});
Object.defineProperty(exports, "MessageOverlay", {
  enumerable: true,
  get: function () {
    return _messageOverlay.default;
  }
});
Object.defineProperty(exports, "Modal", {
  enumerable: true,
  get: function () {
    return _modal.default;
  }
});
Object.defineProperty(exports, "Pagination", {
  enumerable: true,
  get: function () {
    return _paginationLegacy.default;
  }
});
Object.defineProperty(exports, "Password", {
  enumerable: true,
  get: function () {
    return _password.default;
  }
});
Object.defineProperty(exports, "Pivots", {
  enumerable: true,
  get: function () {
    return _pivot.default;
  }
});
Object.defineProperty(exports, "ProgressBar", {
  enumerable: true,
  get: function () {
    return _progressBar.default;
  }
});
Object.defineProperty(exports, "ProgressSteps", {
  enumerable: true,
  get: function () {
    return _progressSteps.default;
  }
});
Object.defineProperty(exports, "QuantitySelector", {
  enumerable: true,
  get: function () {
    return _quantityselector.default;
  }
});
Object.defineProperty(exports, "Radio", {
  enumerable: true,
  get: function () {
    return _radio.default;
  }
});
Object.defineProperty(exports, "Search", {
  enumerable: true,
  get: function () {
    return _search.default;
  }
});
Object.defineProperty(exports, "Spinner", {
  enumerable: true,
  get: function () {
    return _spinner.default;
  }
});
Object.defineProperty(exports, "Table", {
  enumerable: true,
  get: function () {
    return _table.default;
  }
});
Object.defineProperty(exports, "Tabs", {
  enumerable: true,
  get: function () {
    return _tabs.default;
  }
});
Object.defineProperty(exports, "TelephoneInput", {
  enumerable: true,
  get: function () {
    return _telephoneInput.default;
  }
});
Object.defineProperty(exports, "Text", {
  enumerable: true,
  get: function () {
    return _text.default;
  }
});
Object.defineProperty(exports, "Toggle", {
  enumerable: true,
  get: function () {
    return _toggle.default;
  }
});
Object.defineProperty(exports, "Tooltip", {
  enumerable: true,
  get: function () {
    return _tooltipLegacy.default;
  }
});
exports.Validation = void 0;
Object.defineProperty(exports, "request", {
  enumerable: true,
  get: function () {
    return _request.default;
  }
});
exports.versions = exports.utils = void 0;

var _accordion = _interopRequireDefault(require("@ux/accordion"));

var _alert = _interopRequireDefault(require("@ux/alert"));

var _button = _interopRequireDefault(require("@ux/button"));

var _checkbox = _interopRequireDefault(require("@ux/form-element/checkbox"));

var _collapsible = _interopRequireDefault(require("@ux/collapsible"));

var _criteria = _interopRequireDefault(require("@ux/form-element/criteria"));

var _dropdown = _interopRequireDefault(require("@ux/dropdown"));

var _fieldset = _interopRequireDefault(require("@ux/fieldset"));

var _growl = _interopRequireDefault(require("@ux/growl"));

var _messageOverlay = _interopRequireDefault(require("@ux/message-overlay"));

var _modal = _interopRequireDefault(require("@ux/modal"));

var _paginationLegacy = _interopRequireDefault(require("@ux/pagination-legacy"));

var _password = _interopRequireDefault(require("@ux/form-element/password"));

var _pivot = _interopRequireDefault(require("@ux/pivot"));

var _progressBar = _interopRequireDefault(require("@ux/progress-bar"));

var _progressSteps = _interopRequireDefault(require("@ux/progress-steps"));

var _quantityselector = _interopRequireDefault(require("@ux/form-element/quantityselector"));

var _radio = _interopRequireDefault(require("@ux/form-element/radio"));

var _search = _interopRequireDefault(require("@ux/form-element/search"));

var _spinner = _interopRequireDefault(require("@ux/spinner"));

var _table = _interopRequireDefault(require("@ux/table"));

var _tabs = _interopRequireDefault(require("@ux/tabs"));

var _telephoneInput = _interopRequireDefault(require("@ux/form-element/telephone-input"));

var _text = _interopRequireDefault(require("@ux/text"));

var _toggle = _interopRequireDefault(require("@ux/toggle"));

var _tooltipLegacy = _interopRequireDefault(require("@ux/tooltip-legacy"));

var _request = _interopRequireDefault(require("@godaddy/request"));

var utils = _interopRequireWildcard(require("@ux/util"));

exports.utils = utils;

var Validation = _interopRequireWildcard(require("react-validation-context"));

exports.Validation = Validation;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

//
// (Explicit Dependencies) Import UXCore2 components.
//
const versions = {
  // eslint-disable-next-line no-process-env
  [process.env.BUNDLE_NAME]: process.env.BUNDLE_VERSION
}; //
// Due to Safari lacking support for :focus-visible we need to add a temp script
// that adds similar behavior. This script adds a class on the body and using css
// variables we use :focus to do the same styles we would with :focus-visible.
// Once Safari has adequate :focus-visible support this js code will be removed
// and styles will switch from :focus to :focus-visible.
//

exports.versions = versions;
utils.registerKeyboardNavListener(); // Babel7 doesn't attach __esModule onto `import *`s even if they are actually es modules.
// For now, manually attach them for compatibility

[utils].forEach(function (thing) {
  Object.defineProperty(thing, '__esModule', {
    value: true
  });
}); //
// Also export Components to allow other applications depend directly on
// UXCore2. This cannot be done dynamically by destructuring the components as
// this would violate the assumption of statically defined imports/exports.
//
// Note: do NOT export as default, it will prevent `{ [Component] }` like imports.
//
