// @ux/progress-steps/src/progress-step.js
import React from 'react';
import PropTypes from 'prop-types';
import Checkmark from '@ux/icon/checkmark';

/**
 * Possible status types
 *
 * @type {object}
 */
const STATUS = {
  Incomplete: 'incomplete',
  Current: 'current',
  Complete: 'complete'
};

/**
 * ProgressStep component, representing each individual step.
 *
 * @param {Object} props Props
 * @returns {ReactElement} ReactElement
 */
function ProgressStep(props) {

  return (
    <li role='status' aria-live='polite' aria-label={ `step ${props.status}` } className={ `ux-tabs-wiz-step-${props.status}` }>
      <a href={ props.href } onClick={ props.onClick } className={ 'ux-tabs-wiz-link' } tabIndex={ props.href ? null : props.tabIndex }>
        { props.status === 'complete' && <Checkmark fill='white'/> }
        <span className='ux-tabs-wiz-txt'>{ props.children }</span>
      </a>
    </li>
  );
}

ProgressStep.propTypes = {
  status: PropTypes.string.isRequired,
  href: PropTypes.string,
  onClick: PropTypes.func,
  children: PropTypes.node,
  tabIndex: PropTypes.oneOf([-1, 0])
};

ProgressStep.defaultProps = {
  status: STATUS.Incomplete,
  tabIndex: 0
};

ProgressStep.Status = STATUS;

export {
  ProgressStep as default
};
