// @ux/progress-steps/src/styles.js
import { css } from 'linaria';
import intents from './intents.config.js';

export default css`
:global() {
  @import '~@ux/icon/checkmark/index.css';

  .ux-tabs-wiz-list {
    display: flex;
    width: 100%;
    min-height: 2rem;
    padding: 0;
    margin: 0;
    text-align: center;
    list-style: none;
    table-layout: fixed;
    counter-reset: wizList;

    > li {
      position: relative;
      flex: 1;
      display: flex;
      flex-direction: column;
      margin: 0;
      vertical-align: top;
      counter-increment: wizList;
      color: ${intents.uxProgressSteps.item.foregroundColor};
      font-family: ${intents.uxProgressSteps.item.fontFamily};
      font-weight: ${intents.uxProgressSteps.item.fontWeight};
      font-size: ${intents.uxProgressSteps.item.fontSize};
      line-height: ${intents.uxProgressSteps.item.lineHeight};

      /* Lines */
      &:before,
      &:after {
        content: '';
        position: absolute;
        top: .75rem;
        box-sizing: content-box;
        display: block;
        height: .5rem;
        background: currentColor;
        opacity: .3;
      }
      &:before {
        right: 50%;
        left: 0;
        margin-right: calc((2rem / 2) + (.5rem * 2));
      }
      &:after {
        right: 0;
        left: 50%;
        margin-left: calc((2rem / 2) + (.5rem * 2));
      }
      &:first-child:before,
      &:last-child:after {
        display: none;
      }

      /* Clickable area */
      > .ux-tabs-wiz-link {
        position: relative;
        display: flex;
        flex-direction: column;
        justify-content: center;
        align-items: center;
        text-decoration: none;
        color: ${intents.uxProgressSteps.label.foregroundColor};
        opacity: .6;
        &:focus,
        &:focus-visible {
          outline: 0;
          outline: var(--uxp-focus-visible-outline);
        }

        /* Number */
        &:before {
          display: inline-block;
          width: 2rem;
          height: 2rem;
          margin: 0 .5rem;
          line-height: calc(2rem + .2rem);
          text-align: center;
          content: counter(wizList);
          border-radius: 50%;
        }
      }
    }
  }

  .ux-tabs-wiz-txt {
    position: relative;
    display: inline-block;
    max-width: 9rem;
    margin: 0 auto;
    padding-top: calc(${intents.ux.box.density} * 2);
    color: ${intents.uxProgressSteps.label.foregroundColor};
    font-size: ${intents.uxProgressSteps.label.fontSize};
    line-height: ${intents.uxProgressSteps.label.lineHeight};
    font-family: ${intents.uxProgressSteps.label.fontFamily};
    font-weight: ${intents.uxProgressSteps.label.fontWeight};
  }

  .ux-tabs-wiz-list {
    .ux-tabs-wiz-step-complete {
      &:before,
      &:after {
        opacity: 1;
      }
      > .ux-tabs-wiz-link {
        opacity: 1;
        svg {
          display: inline-flex;
          width: 1.25rem;
          height: 2rem;
          position: absolute;
          top: 0;
          left: 50%;
          transform: translateX(-50%);
        }

        &:before {
          background: ${intents.uxProgressSteps.itemCompleted.backgroundColor};
          color: ${intents.uxProgressSteps.itemCompleted.foregroundColor};
          content: '';
        }
      }
    }

    .ux-tabs-wiz-step-current {
      &:before {
        opacity: 1;
      }
    }

    .ux-tabs-wiz-step-current,
    .ux-tabs-wiz-step-incomplete:hover {
      > a {
        opacity: 1;
        &:before {
          color: ${intents.uxProgressSteps.itemHovered.foregroundColor};
          background: ${intents.uxProgressSteps.itemHovered.backgroundColor};
        }
      }
      .ux-tabs-wiz-txt {
        mix-blend-mode: difference;
        font-weight: bolder;
      }
    }
  }
}`;
