// @ux/stylevars/src/mixins/border-radius.js
import { enableRounded, enableRoundedContainers } from '../variables/_variables';

function borderRadius(radius = '4px') {
  if (enableRounded) {
    return `border-radius: ${radius};`;
  }
  return '';
}

function borderRadiusContainers(radius = '4px') {
  if (enableRoundedContainers) {
    return `border-radius: ${radius};`;
  }
  return '';
}

function borderTopLeftRadius(radius) {
  if (enableRounded) {
    return `border-top-left-radius: ${radius};`;
  }
  return '';
}

function borderTopRightRadius(radius) {
  if (enableRounded) {
    return `border-top-right-radius: ${radius};`;
  }
  return '';
}

function borderBottomRightRadius(radius) {
  if (enableRounded) {
    return `border-bottom-right-radius: ${radius};`;
  }
  return '';
}

function borderBottomLeftRadius(radius) {
  if (enableRounded) {
    return `border-bottom-left-radius: ${radius};`;
  }
  return '';
}

function borderTopRadius(radius) {
  return `
    ${borderTopRightRadius(radius)}
    ${borderTopLeftRadius(radius)}
  `;
}

function borderRightRadius(radius) {
  return `
    ${borderBottomRightRadius(radius)}
    ${borderTopRightRadius(radius)}
  `;
}

function borderBottomRadius(radius) {
  return `
    ${borderBottomRightRadius(radius)}
    ${borderBottomLeftRadius(radius)}
  `;
}

function borderLeftRadius(radius) {
  return `
    ${borderBottomLeftRadius(radius)}
    ${borderTopLeftRadius(radius)}
  `;
}

export {
  borderRadius,
  borderRadiusContainers,
  borderTopLeftRadius,
  borderTopRadius,
  borderTopRightRadius,
  borderBottomLeftRadius,
  borderBottomRightRadius,
  borderLeftRadius,
  borderBottomRadius,
  borderRightRadius
};
