// @ux/stylevars/src/mixins/breakpoints.js
import { stripUnit, math } from 'polished';
import { gridBreakpoints } from '../variables/_variables';

function breakpointNext(name, breakpoints) {
  const breakpointNames = Object.keys(breakpoints);
  const n = breakpointNames.indexOf(name);
  return n < (breakpointNames.length - 1) ? breakpointNames[n + 1] : null;
}

function breakpointMin(name, breakpoints) {
  const min = parseInt(stripUnit(breakpoints[name]), 10);
  return min !== 0 ? breakpoints[name] : null;
}

function breakpointMax(name, breakpoints) {
  const next = breakpointNext(name, breakpoints);
  return next ? math(`${breakpointMin(next, breakpoints)} - 1px`) : null;
}

function breakpointInfix(name, breakpoints) {
  return breakpointMin(name, breakpoints) == null ? '' : `-${name}`;
}

function mediaBreakpointUp(name, content) {
  const min = breakpointMin(name, gridBreakpoints);
  if (min) {
    return `
      @media (min-width: ${min}) {
        ${content}
      }
    `;
  }
  return `${content}`;
}

function mediaBreakpointDown(name, content) {
  const max = breakpointMax(name, gridBreakpoints);
  if (max) {
    return `
      @media (max-width: ${max}) {
        ${content}
      }
    `;
  }
  return `${content}`;
}

export {
  breakpointInfix,
  breakpointMax,
  breakpointMin,
  breakpointNext,
  mediaBreakpointDown,
  mediaBreakpointUp
};
