// @ux/stylevars/src/mixins/gradients.js
/* eslint-disable max-len */
import { mix, rgba } from 'polished';
import { enableGradients, bodyBg } from '../variables/_variables';

function gradientBg(color) {
  if (enableGradients) {
    return `background: ${color} linear-gradient(180deg, ${mix(0.15, bodyBg, color)} repeat-x`;
  }
  return `background-color: ${color}`;
}

function gradientY(startColor = '#555', endColor = '#333', startPercent = '0%', endPercent = '100%') {
  return `
    background-image: linear-gradient(to bottom, ${startColor}, ${startPercent}, ${endColor}, ${endPercent});
    backgrond-repeat: repeat-x;
  `;
}

function gradientDirectional(startColor = '#555', endColor = '#333', deg = '45deg') {
  return `
    background-image: linear-gradient(${deg}, ${startColor}, ${endColor});
    background-repeat: repeat-x;
  `;
}

function gradientXThreeColors(startColor = '#00b3ee', midColor = '#7a43b6', colorStop = '50%', endColor = '#c3325f') {
  return `
    background-image: linear-gradient(to right, ${startColor}, ${midColor} ${colorStop}, ${endColor});
    background-repeat: no-repeat;
  `;
}

function gradientYThreeColors(startColor = '#00b3ee', midColor = '#7a43b6', colorStop = '50%', endColor = '#c3325f') {
  return `
    background-image: linear-gradient(${startColor}, ${midColor} ${colorStop}, ${endColor});
    background-repeat: no-repeat;
  `;
}

function gradientRadial(innerColor = '#555', outerColor = '#333') {
  return `
    background-image: radial-gradient(circle, ${innerColor}, ${outerColor});
    background-repeat: no-repeat;
  `;
}

function gradientStriped(color = rgba(255, 255, 255, 0.15), angle = '45deg') {
  // eslint-disable-next-line max-len
  return `background-image: linear-gradient(${angle}, ${color} 25%, transparent 25%, transparent 50%, ${color} 50%, ${color} 75%, transparent 75%, transparent);`;
}

export {
  gradientBg,
  gradientDirectional,
  gradientRadial,
  gradientStriped,
  gradientXThreeColors,
  gradientYThreeColors,
  gradientY
};
