// @ux/stylevars/src/mixins/shadows.js
/* eslint-disable max-params */
import { rgba } from 'polished';
import * as vars from '../variables/_variables';
import palette from '../functions/palette';

export function boxShadow(shadow) {
  if (vars.enableShadows) {
    return `box-shadow: ${shadow}`;
  }
  return '';
}

export function mrShadow(size = '4px', opacity = 0.1, dir = 'default', color = palette('black', 'dark')) {
  let sizeH = size;
  if (dir === 'down') {
    sizeH = '0';
  }
  return boxShadow(`${sizeH} ${size} 0 0 ${rgba(color, opacity)}`);
}

export function shadowX(size, opacity, blur, spread, color = palette('gray')) {
  const sizeH = '0';
  return boxShadow(`${sizeH} ${size} ${blur} ${spread} ${rgba(color, opacity)}`);
}

export function shadow1() {
  const { shadow1Size, shadow1Opacity, shadow1Blur, shadow1Spread } = vars;
  return shadowX(shadow1Size, shadow1Opacity, shadow1Blur, shadow1Spread);
}

export function shadow2() {
  const { shadow2Size, shadow2Opacity, shadow2Blur, shadow2Spread } = vars;
  return shadowX(shadow2Size, shadow2Opacity, shadow2Blur, shadow2Spread);
}

export function shadow3() {
  const { shadow3Size, shadow3Opacity, shadow3Blur, shadow3Spread } = vars;
  return shadowX(shadow3Size, shadow3Opacity, shadow3Blur, shadow3Spread);
}

export function shadow4() {
  const { shadow4Size, shadow4Opacity, shadow4Blur, shadow4Spread } = vars;
  return shadowX(shadow4Size, shadow4Opacity, shadow4Blur, shadow4Spread);
}
