// @ux/stylevars/src/mixins/types.js
import { marketFont } from '../functions/markets';
import palette from '../functions/palette';

// Used to create a bottom margin that is based on the heading's font-size and base our line-height.
function heading(size, height) {
  return `
    font-size: ${size};
    line-height: ${height};
  `;
}

// Used in Dropdowns, Growls, Tippers
function barHeader() {
  return `
    position: relative;
    color: ${palette('black')};
  `;
}

function fontBase() {
  return `font-family: ${marketFont('base')};`;
}

function fontPrimaryBold() {
  return `
    font-family: ${marketFont('primary-bold', 'family')};
    font-weight: ${marketFont('primary-bold', 'weight')};
  `;
}

// gd-sage only
function fontMarketing() {
  return `
    font-family: ${marketFont('marketing', 'family')};
    font-weight: ${marketFont('marketing', 'weight')} !important;
  `;
}

// GD-Sage
function fontSage() {
  return `
    font-family: ${marketFont('sage', 'family')};
    font-weight: ${marketFont('sage', 'weight')};
  `;
}

function fontSageBold() {
  return `
    font-family: ${marketFont('sage-bold', 'family')};
    font-weight: ${marketFont('sage-bold', 'weight')};
  `;
}

// Used for .headline-primary and form things.
function menuTitle() {
  return fontPrimaryBold();
}

export {
  heading,
  menuTitle,
  barHeader,
  fontMarketing,
  fontBase,
  fontPrimaryBold,
  fontSage,
  fontSageBold
};
