// @ux/progress-bar/src/styles.js
import { css } from 'linaria';
import intents from './intents.config.js';

function progressColorVariant(backgroundColor, foregroundColor) {
  return `
    .ux-progress-fill {
      background-color: ${backgroundColor};
      .ux-progress-text {
        color: ${foregroundColor};
      }
    }
    circle.ux-progress-fill {
      stroke: ${backgroundColor};
    }
  `;
}

function gradientStriped(alpha = 0.1, rotation = '45deg') {
  const color = `rgba(0, 0, 0, ${alpha})`;
  return `background-image: linear-gradient(${rotation}, ${color} 25%, transparent 25%, transparent 50%, ${color} 50%, ${color} 75%, transparent 75%, transparent);`;
}
const CIRCLE_SIZE_XS = '3.25rem';
const CIRCLE_SIZE_SM = '6.25rem';
const CIRCLE_SIZE_MD = '9.5rem';
const CIRCLE_SIZE_LG = '15.5rem';

export default css`
:global() {
  .ux-progress {
    display: flex;
    flex-direction: column;
  }

  /* Title */
  .ux-progress-label {
    margin-bottom: 5px;
    display: inline-flex;
    font-family: ${intents.uxProgressBar.title.fontFamily};
    font-size: ${intents.uxProgressBar.title.fontSize};
    font-weight: ${intents.uxProgressBar.title.fontWeight};
    line-height: ${intents.uxProgressBar.title.lineHeight};
    color: ${intents.uxProgressBar.title.foregroundColor};
  }

  .ux-progress-bar-notice {
    font-family: ${intents.uxProgressBar.description.fontFamily};
    font-size: ${intents.uxProgressBar.description.fontSize};
    font-weight: ${intents.uxProgressBar.description.fontWeight};
    line-height: ${intents.uxProgressBar.description.lineHeight};
    color: ${intents.uxProgressBar.description.foregroundColor};
  }

  /* Text inside the bar */
  .ux-progress-text {
    margin-right: 10px;
    display: flex;
    align-items: center;
    font-family: ${intents.uxProgressBar.meter.fontFamily};
    font-size: ${intents.uxProgressBar.meter.fontSize};
    font-weight: ${intents.uxProgressBar.meter.fontWeight};
    line-height: ${intents.uxProgressBar.meter.lineHeight};
    color: ${intents.uxProgressBar.meter.foregroundColor};
  }

  .ux-progress-bar {
    &.ux-progress-striped {
      .ux-progress-fill {
        animation: progress-bar-stripes 500ms linear infinite;
        background-size: 1.5rem 1.5rem;
        ${gradientStriped()};
      }
    }
    &.ux-progress-tick-marks {
      .ux-progress-area {
        position: relative;
      }
      .ux-progress-tick-labels {
        display: flex;
        justify-content: space-between;
      }
      .ux-progress-tick-label {
        flex: 1;
        height: 1.5rem;
        font-family: ${intents.uxProgressBar.item.fontFamily};
        font-size: ${intents.uxProgressBar.item.fontSize};
        font-weight: ${intents.uxProgressBar.item.fontWeight};
        line-height: ${intents.uxProgressBar.item.lineHeight};
        color: ${intents.uxProgressBar.item.foregroundColor};
        span {
          position: absolute;
          transform: translateX(-50%);
        }
      }
      .ux-progress-tick-label:last-of-type {
        flex: 0;
      }
      .ux-progress-tick-lines {
        display: flex;
        justify-content: space-evenly;
        position: absolute;
        top: 0;
        right: 0;
        bottom: 0;
        left: 0;
      }
      .ux-progress-tick-line {
        background-color: ${intents.uxProgressBar.item.backgroundColor};
      }
    }

    .ux-progress-area {
      display: flex;
      width: 100%;
      background-color: ${intents.uxProgressBar.backgroundColor};
      overflow: hidden;
      border-radius: ${intents.uxProgressBar.borderRadius};
    }
    .ux-progress-fill {
      display: flex;
      justify-content: flex-end;
      height: 100%;
      background-color: ${intents.uxProgressBar.meter.backgroundColor};
    }
    .ux-progress-text-start {
      color: ${intents.ux.boxLowContrast.foregroundColor};
      margin-right: -30px;
    }

    &.ux-progress-sm {
      .ux-progress-text {
        margin-top: 2px;
        display: flex;
        float: none !important;
        color: inherit;
      }
    }
  }

  .ux-progress-circle {
    position: relative;
    align-items: center;
    svg,
    label,
    small {
      display: flex;
    }
    .ux-progress-label,
    small {
      justify-content: center;
      text-align: center;
      width: 100%;
    }
    circle {
      stroke: ${intents.uxProgressBar.backgroundColor};
    }
    circle.ux-progress-fill {
      stroke: ${intents.uxProgressBar.meter.backgroundColor};
    }
    .ux-progress-text {
      margin: 0;
      display: flex;
      justify-content: center;
      align-items: center;
      position: absolute;
      top: 0;
      left: 0;
      color: inherit;
    }
  }

  .ux-progress-bar {
    &.ux-progress-md,
    &.ux-progress-lg {
      .ux-progress-area {
        height: 1.5rem;
      }
    }
    &.ux-progress-sm {
      .ux-progress-area {
        height: .75rem;
      }
    }
    &.ux-progress-xs {
      .ux-progress-area {
        height: .375rem;
      }
    }
  }

  .ux-progress-circle {
    &.ux-progress-xs {
      width: ${CIRCLE_SIZE_XS};
      font-size: 80%;
      .ux-progress-text {
        width: ${CIRCLE_SIZE_XS};
        height: ${CIRCLE_SIZE_XS};
        font-size: calc(${CIRCLE_SIZE_XS} / 4);
      }
    }
    &.ux-progress-sm {
      width: ${CIRCLE_SIZE_SM};
      .ux-progress-text {
        width: ${CIRCLE_SIZE_SM};
        height: ${CIRCLE_SIZE_SM};
        font-size: calc(${CIRCLE_SIZE_SM} / 4);
      }
    }
    &.ux-progress-md {
      width: ${CIRCLE_SIZE_MD};
      .ux-progress-text {
        width: ${CIRCLE_SIZE_MD};
        height: ${CIRCLE_SIZE_MD};
        font-size: calc(${CIRCLE_SIZE_MD} / 4);
      }
    }
    &.ux-progress-lg {
      width: ${CIRCLE_SIZE_LG};
      .ux-progress-text {
        width: ${CIRCLE_SIZE_LG};
        height: ${CIRCLE_SIZE_LG};
        font-size: calc(${CIRCLE_SIZE_LG} / 4);
      }
    }
  }

  .ux-progress-success {
    ${progressColorVariant(
    intents.uxProgressBar.meterSuccess.backgroundColor,
    intents.uxProgressBar.meterSuccess.foregroundColor
  )}
  }

  .ux-progress-info {
    ${progressColorVariant(
    intents.uxProgressBar.meterInfo.backgroundColor,
    intents.uxProgressBar.meterInfo.foregroundColor
  )}
  }

  .ux-progress-warning {
    ${progressColorVariant(
    intents.uxProgressBar.meterWarning.backgroundColor,
    intents.uxProgressBar.meterWarning.foregroundColor
  )}
  }

  .ux-progress-danger {
    ${progressColorVariant(
    intents.uxProgressBar.meterCritical.backgroundColor,
    intents.uxProgressBar.meterCritical.foregroundColor
  )}
  }

  @keyframes progress-bar-stripes {
    0%    { background-position: 1.5rem 0; }
    100%  { background-position: 0 0; }
  }
}
`;
