// @ux/form-element/examples/search-event-handlers.stories.js
import { Search } from '@ux/uxcore2';
import React, { Fragment } from 'react';

export default class SearchEventHandlers extends React.Component {

  constructor() {
    super(...arguments);

    this.onSearch = this.onSearch.bind(this);
    this.onChange = this.onChange.bind(this);
    this.onSelect = this.onSelect.bind(this);

    this.state = {
      message: ''
    };
  }

  onSearch(searchTerm) {
    this.setState({ message: `Search Event Fired. Searched term: '${searchTerm}'` });
  }

  onChange(value) {
    this.setState({ message: `Change Event Fired. New value: '${value}'` });
  }

  onSelect(selectedItem) {
    this.setState({ message: `Select Event Fired. Selected item: '${selectedItem}'` });
  }

  render() {
    const data = [
      'Apples',
      'Bananas',
      'Blueberries',
      'Blackberries',
      'Grapes',
      'Peaches',
      'Snozberries',
      'Yams'
    ];

    return (
      <Fragment>
        { this.state.message }
        <Search
          data={ data }
          maxResults={ 10 }
          onSearch={ this.onSearch }
          onChange={ this.onChange }
          onSelect={ this.onSelect }
          id='eventHandlers'
          placeholder='Type a search term and press enter. Or, select from the items.'
        />
      </Fragment>
    );
  }
}
