// @ux/table/lib/tfoot.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _collection = _interopRequireDefault(require("./collection"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _tr = _interopRequireDefault(require("./tr"));

//
// Set of allowed child Components for the Table Body.
//
const COMPONENTS = {
  tr: _tr.default
};
/**
 * A table footer component
 *
 * @displayName Tfoot
 * @export
 * @class Tfoot
 * @extends {Collection}
 */

class Tfoot extends _collection.default {
  /**
   * Getter that returns the allowed children.
   *
   * @returns {Object} Components
   * @private
   */
  get components() {
    return COMPONENTS;
  }
  /**
   * Render a table footer.
   *
   * @returns {ReactElement} Table footer.
   * @public
   */


  render() {
    const {
      style,
      children,
      className
    } = this.props;
    return /*#__PURE__*/_react.default.createElement("tfoot", {
      style: style,
      className: className
    }, _react.default.Children.map(children, this.renderChildren, this));
  }

}

exports.default = Tfoot;
Tfoot.propTypes = {
  /** Child row */
  children: _propTypes.default.node,

  /** Styles to apply to the element */
  style: _propTypes.default.object,

  /** ClassName to apply to the element */
  className: _propTypes.default.string
};
