// @ux/table/examples/dynamic.stories.js
import React, { useState } from 'react';
import Table from '../src';
import Dropdown from '@ux/dropdown';
const { DropdownItem } = Dropdown;

import '@ux/dropdown/src/styles';

function capitalize(str) {
  return `${str[0].toUpperCase()}${str.slice(1)}`;
}

const characters = [
  { species: 'Wookie', character: 'Chewbacca',   actor: 'Peter Mayhew' },
  { species: 'Human',  character: 'Han Solo',    actor: 'Harrison Ford' },
  { species: 'Droid',  character: 'C-3PO',       actor: 'Anthony Daniels' },
  { species: 'Human',  character: 'Darth Vader', actor: 'James Earl Jones (voice)' },
  { species: 'Human',  character: 'Darth Vader', actor: 'David Prowse' }
];

const scientists = [
  { name: 'Albert Einstein',     born: 1879, died: 1955, fields: ['Physics'] },
  { name: 'Isaac Newton',        born: 1643, died: 1727, fields: ['Physics', 'Astronomy'] },
  { name: 'Marie Curie',         born: 1867, died: 1934, fields: ['Physics', 'Chemistry'] },
  { name: 'Nicolaus Copernicus', born: 1473, died: 1543, fields: ['Astronomy'] },
  { name: 'Charles Darwin',      born: 1809, died: 1882, fields: ['Biology', 'Geology'] }
];

const data = {
  Characters: {
    rows: characters,
    order: ['character', 'species', 'actor']
  },
  Scientists: {
    rows: scientists.map(item => {
      return {
        ...item,
        fields: item.fields.join(', ')
      };
    }),
    order: ['name', 'fields', 'born', 'died']
  }
};

export default function DynamicData() {
  const [selected, setSelected] = useState('Scientists');
  const { order, rows } = data[selected];
  return (
    <div>

      <Dropdown onChange={ (ev) => setSelected(ev.children) }
        id='dropdown'
        defaultSelected={ 1 }
        type='select'>
        { Object.keys(data).map(name => <DropdownItem key={ name }>{ name }</DropdownItem>) }
      </Dropdown>

      <Table
        className='table'
        order={ order }
        data={ rows  }>
        <thead>
          <tr>
            { order.map(col => <th key={ col }>{ capitalize(col) }</th>)}
          </tr>
        </thead>
      </Table>
    </div>
  );
}
