// @ux/form-group/examples/index.stories.js
import { Boilerplate } from '@ux/gxsys-page';
import React from 'react';
import packageJson from '../package.json';

export default class SandboxFormGroup extends Boilerplate {
  introduction() {
    return (
      <p>
        The <code>FormGroup</code> Component is a base Component used by
        <code>@ux/radio-group</code> and <code>@ux/checkbox-group</code> mostly
        similar between both components.
      </p>
    );
  }

  get packageInfo() {
    return {
      source: {
        name: packageJson.name,
        link: packageJson.homepage
      },
      version: packageJson.version,
      installCmd: `npm install ${packageJson.name}`,
      platformSupport: {
        android: true,
        ios: true,
        web: true
      },
      license: packageJson.license || 'Proprietary',
      relatedComponents: [
        'FormElement',
        'CheckboxGroup',
        'RadioGroup'
      ]
    };
  }

  get documentation() {
    return [
      require('./api/src/index.json')
    ];
  }

  get examples() {
    return [];
  }
}

SandboxFormGroup.defaultProps = {
  component: '@ux/form-group',
  documentation: [{
    source: {
      '@ux/form-group': require('!!@ux/doc-generator!../')
    },
    target: '@ux/form-group'
  }],
  // mode: 'gxsys2',
  name: 'Form Group'
};
