// @ux/stylevars/lib/variables/pl/colors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _default = {
  'white': {
    base: {
      color: '#fff',
      contrast: '#111'
    }
  },
  'black': {
    base: {
      color: '#000',
      contrast: '#fff'
    },
    dark: {
      color: '#000',
      contrast: '#fff'
    }
  },
  'gray': {
    base: {
      color: '#999',
      contrast: '#fff'
    },
    faint: {
      color: '#fbfaf8',
      contrast: '#111'
    },
    light: {
      color: '#aea99c',
      contrast: '#111'
    },
    midrange: {
      color: '#555',
      contrast: '#fff'
    },
    dark: {
      color: '#333',
      contrast: '#fff'
    }
  },
  'product': {
    base: {
      color: '#f5f7f8',
      contrast: '#111'
    },
    faint: {
      color: '#e8eae8',
      contrast: '#111'
    },
    light: {
      color: '#d6d6d6',
      contrast: '#111'
    },
    midrange: {
      color: '#bfbfbf',
      contrast: '#111'
    },
    dark: {
      color: '#999',
      contrast: '#111'
    }
  },
  // Blues
  'primary': {
    base: {
      color: '#1d6ccd',
      contrast: '#fff'
    },
    pastel: {
      color: '#dde9f8',
      contrast: '#111'
    },
    highlight: {
      color: '#438be4',
      contrast: '#fff'
    },
    midrange: {
      color: '#1757a5',
      contrast: '#fff'
    },
    dark: {
      color: '#124481',
      contrast: '#fff'
    }
  },
  'primary-o': {
    base: {
      color: '#3690ff',
      contrast: '#111'
    },
    pastel: {
      color: '#e1eeff',
      contrast: '#111'
    },
    highlight: {
      color: '#73b2ff',
      contrast: '#111'
    },
    midrange: {
      color: '#0877ff',
      contrast: '#111'
    },
    dark: {
      color: '#0064de',
      contrast: '#fff'
    }
  },
  'nav': {
    base: {
      color: '#0862cf',
      contrast: '#fff'
    },
    pastel: {
      color: '#dae7f8',
      contrast: '#111'
    },
    highlight: {
      color: '#1e80f6',
      contrast: '#fff'
    },
    midrange: {
      color: '#064da3',
      contrast: '#fff'
    },
    dark: {
      color: '#053a7c',
      contrast: '#fff'
    }
  },
  // Yellows
  'secondary': {
    base: {
      color: '#c62828',
      contrast: '#fff'
    },
    pastel: {
      color: '#f6dfdf',
      contrast: '#111'
    },
    highlight: {
      color: '#dc5050',
      contrast: '#fff'
    },
    midrange: {
      color: '#a02020',
      contrast: '#fff'
    },
    dark: {
      color: '#7e1919',
      contrast: '#fff'
    }
  },
  'secondary-o': {
    base: {
      color: '#c62828',
      contrast: '#111'
    },
    pastel: {
      color: '#f6dfdf',
      contrast: '#111'
    },
    highlight: {
      color: '#dc5050',
      contrast: '#111'
    },
    midrange: {
      color: '#a02020',
      contrast: '#fff'
    },
    dark: {
      color: '#7e1919',
      contrast: '#fff'
    }
  },
  'warning': {
    base: {
      color: '#fedc45',
      contrast: '#111'
    },
    pastel: {
      color: '#fffae3',
      contrast: '#111'
    },
    highlight: {
      color: '#fff44d',
      contrast: '#111'
    },
    midrange: {
      color: '#fed12d',
      contrast: '#111'
    },
    dark: {
      color: '#fec419',
      contrast: '#111'
    }
  },
  // Seafoams
  'info': {
    base: {
      color: '#47cabb',
      contrast: '#111'
    },
    pastel: {
      color: '#e3f7f5',
      contrast: '#111'
    },
    highlight: {
      color: '#77d8cd',
      contrast: '#111'
    },
    midrange: {
      color: '#33b0a2',
      contrast: '#111'
    },
    dark: {
      color: '#2a9185',
      contrast: '#111'
    }
  },
  // Reds
  'danger': {
    base: {
      color: '#db1802',
      contrast: '#fff'
    },
    pastel: {
      color: '#fadcd9',
      contrast: '#111'
    },
    highlight: {
      color: '#fd341d',
      contrast: '#fff'
    },
    midrange: {
      color: '#ae1302',
      contrast: '#fff'
    },
    dark: {
      color: '#850f01',
      contrast: '#fff'
    }
  },
  // Greens
  'success': {
    base: {
      color: '#419a18',
      contrast: '#fff'
    },
    pastel: {
      color: '#e3f0dc',
      contrast: '#111'
    },
    highlight: {
      color: '#57cf20',
      contrast: '#111'
    },
    midrange: {
      color: '#307212',
      contrast: '#fff'
    },
    dark: {
      color: '#214f0c',
      contrast: '#fff'
    }
  },
  // Magentas - feature
  'feature': {
    base: {
      color: '#e20087',
      contrast: '#fff'
    },
    pastel: {
      color: '#fbd9ed',
      contrast: '#111'
    },
    highlight: {
      color: '#ff20a5',
      contrast: '#111'
    },
    midrange: {
      color: '#b4006c',
      contrast: '#fff'
    },
    dark: {
      color: '#8b0053',
      contrast: '#fff'
    }
  },
  // Naturals
  'natural': {
    base: {
      color: '#fbb684',
      contrast: '#111'
    },
    pastel: {
      color: '#ffe8c9',
      contrast: '#111'
    },
    highlight: {
      color: '#fed9b4',
      contrast: '#111'
    },
    midrange: {
      color: '#bd8150',
      contrast: '#111'
    },
    dark: {
      color: '#a36841',
      contrast: '#fff'
    }
  }
};
exports.default = _default;
