// @ux/dropdown/examples/select.stories.js
/* eslint no-console: 0 */
import React, { Fragment, useState, useCallback } from 'react';
import Button from '@ux/button';
import '@ux/button/styles';
import Dropdown from '@ux/dropdown';
import '@ux/dropdown/styles';

const { DropdownItem } = Dropdown;
const cities = {
  nyc: 'New York',
  sfo: 'San Francisco',
  phx: 'Phoenix',
  ams: 'Amsterdam',
  brl: 'Berlin',
  lon: 'London',
  hkg: 'Hong Kong',
  bei: 'Beijing',
  gro: 'Groningen'
};

export default function Select() {
  const [selected, setSelected] = useState([]);

  const changed = useCallback((props) => {
    console.log('Dropdown properties changed', props);
  });

  const select = useCallback(() => {
    setSelected([0, 1, 5]);
  });
  return (
    <Fragment>
      <Dropdown type='multiselect' name='myDropdown' change={ changed } selected={ selected }>
        {Object.keys(cities).map(city => {
          return <DropdownItem key={ city } value={ city }>{cities[city]}</DropdownItem>;
        })}
      </Dropdown>
      <Button design='primary' type='submit' id='city-select-button' onClick={ select } text='Select NYC, SFO, and LON' />
    </Fragment>
  );
}
