// @ux/spinner/src/styles.js
import { css } from 'linaria';
import intents from './intents.config.js';

export default css`
:global() {
  .ux-spinner {
    transform: rotate(270deg);
    width: 2.25em;
    height: 2.25em;
    margin: 0 .25rem;
    align-self: center;
    &.ux-spinner-sm {
      width: 1.25em;
      height: 1.25em;
    }
    &.ux-spinner-lg {
      width: 4em;
      height: 4em;
    }
  }


  .bg-spinner {
    stroke: ${intents.uxSpinner.backgroundColor};
    stroke-width: .6em;
  }

  /* Circle styles. */

  .fg-spinner {
    animation: 1.9s linear infinite both circle-animation;
    display: block;
    fill: transparent;
    stroke-linecap: round;
    stroke: ${intents.uxSpinner.foregroundColor};
    stroke-dasharray: 195;
    stroke-dashoffset: 192;
    stroke-width: .6em ;
    transform-origin: 50% 50%;
  }

  /* Circle animation. */
  @keyframes circle-animation {
    0% {
      stroke-dashoffset: 192;
      transform: rotate(0);
    }

    50% {
      stroke-dashoffset: 96;
      transform: rotate(180deg);
    }

    100% {
      stroke-dashoffset: 192;
      transform: rotate(360deg);
    }
  }
}
`;
