// @ux/tooltip-legacy/src/styles.js
import { css } from 'linaria';
import intents from './intents.config.js';
import '@ux/button/src/styles';

export default css`
:global() {
  @import "~@ux/icon/x/index.css";
  @import "~@ux/icon/help/index.css";

  .tooltip-help-icon,
  .tooltip-over-text {
    &:focus {
      outline: var(--uxp-focus-visible-outline, none);
    }
  }

  /* Base class */
  .tooltip {
    font-family: ${intents.uxTooltip.fontFamily};
    font-weight: ${intents.uxTooltip.fontWeight};
    line-height: ${intents.uxTooltip.lineHeight};
    font-size: ${intents.uxTooltip.fontSize};
    padding: .5rem;
    z-index: 1070;
    position: absolute;
    display: block;
    max-width: 22rem;
    opacity: 0;
    margin: 0;

    /* Allow breaking very long words so they don't overflow the tooltip's bounds */
    word-wrap: break-word;

    a {
      color: inherit;
      text-decoration: underline;
      &:hover {
        color: inherit;
      }
    }

    &.show { opacity: 1; }

    .tooltip-arrow {
      width: 9px;
      height: 9px;
      position: absolute;
      display: block;
      border-color: transparent !important;
      border-width: 0 10px 10px !important;
      border-bottom-color: ${intents.uxTooltip.borderColor} !important;

      &::before {
        position: absolute;
        content: "";
        border-color: transparent;
        border-style: solid;
      }

      &:after {
        position: absolute;
        top: 1px;
        left: -9px;
        display: block;
        width: 0;
        height: 0;
        content: "";
        border-color: transparent;
        border-style: solid;
        border-width: 0 9px 9px;
        border-bottom-color: ${intents.uxTooltip.backgroundColor};
      }
    }

    .tooltip-close {
      background: transparent;
      float: right;
      color: inherit;
      cursor: pointer;
      position: absolute;
      top: 0;
      right: 0;

      &:focus {
        outline: var(--uxp-focus-visible-outline, none);
      }
    }

    &.bs-tooltip-top,
    &.bs-tooltip-top-right,
    &.bs-tooltip-top-left {
      margin-top: .25rem;
      .tooltip-arrow {
        top: 0;
        left: 50%;
      }
    }

    /* Bottom */
    &.bs-tooltip-bottom,
    &.bs-tooltip-bottom-right,
    &.bs-tooltip-bottom-left {
      margin-top: -.25rem;
      .tooltip-arrow {
        bottom: 0;
        left: 50%;
        transform: rotate(180deg);
        &:after {
          left: 0;
        }
      }
    }
  }

  /* Body that contains bg color */
  .tooltip-inner {
    border-radius: ${intents.uxTooltip.borderRadius};
    position: relative;
    background-color: ${intents.uxTooltip.backgroundColor};
    color: ${intents.uxTooltip.foregroundColor};
    padding: 1.25rem;
    text-align: left;
    text-decoration: none;
    word-wrap: break-word;
    border: 1px solid ${intents.uxTooltip.borderColor};
    max-width: none;
  }

  /* Titles */
  .tooltip-title {
    margin-bottom: 1rem;
    font-family: ${intents.uxTooltip.title.fontFamily};
    font-size: ${intents.uxTooltip.title.fontSize};
    font-weight: ${intents.uxTooltip.title.fontWeight};
    line-height: ${intents.uxTooltip.title.lineHeight};
  }
}`;
