// @ux/dropdown/src/intents.config.js
import Intents, { enums } from '@ux/intents';

// eslint-disable-next-line max-statements
export default (function () {
  const intents = new Intents();

  intents.create('@ux/dropdown', {
    // basic .dropdown-menu
    styleProperties: [
      enums.properties.FONT_FAMILY,
      enums.properties.FONT_SIZE,
      enums.properties.LINE_HEIGHT,
      enums.properties.FOREGROUND_COLOR,
      enums.properties.BACKGROUND_COLOR,
      enums.properties.BORDER_COLOR,
      enums.properties.BORDER_RADIUS
    ],
    expressions: [
      enums.expressions.CHOSEN
    ],
    features: {
      // .dropdown-item
      [enums.features.ITEM]: {
        styleProperties: [
          enums.properties.FONT_FAMILY,
          enums.properties.FONT_SIZE,
          enums.properties.FONT_WEIGHT,
          enums.properties.LINE_HEIGHT,
          enums.properties.FOREGROUND_COLOR,
          enums.properties.BACKGROUND_COLOR
        ],
        interactions: [
          enums.interactions.HOVERED
        ]
      },
      [enums.features.INDICATOR]: {
        styleProperties: [
          enums.properties.BORDER_COLOR
        ]
      },
      // .dropdown-header in dropdowns
      [enums.features.TITLE]: {
        styleProperties: [
          enums.properties.FONT_FAMILY,
          enums.properties.FONT_SIZE,
          enums.properties.FONT_WEIGHT,
          enums.properties.LINE_HEIGHT,
          enums.properties.FOREGROUND_COLOR,
          enums.properties.BACKGROUND_COLOR
        ]
      },
      [enums.features.BACKDROP]: {
        styleProperties: [
          enums.properties.BACKGROUND_COLOR
        ]
      },
      formControl: {
        styleProperties: [
          enums.properties.BORDER_COLOR
        ],
        expressions: [
          enums.expressions.CRITICAL
        ]
      },
      placeholder: {
        styleProperties: [
          enums.properties.FOREGROUND_COLOR
        ]
      },
      tripleDotOpen: {
        styleproperties: [
          enums.properties.FOREGROUND_COLOR,
          enums.properties.BACKGROUND_COLOR
        ]
      }
    }
  });

  // .dropdown-menu styles
  intents.uxDropdown.fontFamily = intents.ux.text.fontFamily;
  intents.uxDropdown.fontSize = intents.ux.text.fontSize;
  intents.uxDropdown.lineHeight = intents.ux.text.lineHeight;
  intents.uxDropdown.foregroundColor = intents.ux.box.foregroundColor;
  intents.uxDropdown.backgroundColor = intents.ux.box.backgroundColor;
  intents.uxDropdown.borderColor = intents.ux.control.outlineColor; // Expected to show as "focused"
  intents.uxDropdown.borderRadius = intents.ux.box.borderRadius;

  // Backdrop for mobile
  intents.uxDropdown.backdrop.backgroundColor = intents.ux.boxHighContrastOverlay.backgroundColor;

  // .tripledot things
  intents.uxDropdownChosen.foregroundColor = intents.ux.feedback.onFeedbackColor;
  intents.uxDropdownChosen.backgroundColor = intents.ux.feedback.feedbackColor;

  // .dropdown-item
  intents.uxDropdown.item.fontFamily = intents.ux.textAction.fontFamily;
  intents.uxDropdown.item.fontSize = intents.ux.textAction.fontSize;
  intents.uxDropdown.item.fontWeight = intents.ux.textAction.fontWeight;
  intents.uxDropdown.item.lineHeight = intents.ux.textAction.lineHeight;
  intents.uxDropdown.item.foregroundColor = intents.ux.action.foregroundColor;
  intents.uxDropdown.item.backgroundColor = intents.ux.action.backgroundColor;

  // .dropdown-item hover
  intents.uxDropdown.itemHovered.foregroundColor = intents.ux.actionHovered.foregroundColor;
  intents.uxDropdown.itemHovered.backgroundColor = intents.ux.actionHovered.backgroundColor;

  // .dropdown-divider border color
  intents.uxDropdown.indicator.borderColor = intents.ux.box.borderColor;

  // .dropdown-menu .form-control
  intents.uxDropdown.formControl.borderColor = intents.ux.control.borderColor;
  intents.uxDropdown.formControlCritical.borderColor = intents.ux.feedbackHighContrastCritical.feedbackColor;

  // .dropdown placeholder
  intents.uxDropdown.placeholder.foregroundColor = intents.ux.controlVoid.foregroundColor;

  // .dropdown-header
  intents.uxDropdown.title.fontFamily = intents.ux.textLabel.fontFamily;
  intents.uxDropdown.title.fontSize = intents.ux.textLabel.fontSize;
  intents.uxDropdown.title.fontWeight = intents.ux.textLabel.fontWeight;
  intents.uxDropdown.title.lineHeight = intents.ux.textLabel.lineHeight;
  intents.uxDropdown.title.foregroundColor = intents.ux.box.foregroundColor;
  intents.uxDropdown.title.backgroundColor = intents.ux.box.backgroundColor;

  return intents;
}());
