// @ux/stylevars/lib/mixins/border-radius.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.borderBottomLeftRadius = borderBottomLeftRadius;
exports.borderBottomRadius = borderBottomRadius;
exports.borderBottomRightRadius = borderBottomRightRadius;
exports.borderLeftRadius = borderLeftRadius;
exports.borderRadius = borderRadius;
exports.borderRadiusContainers = borderRadiusContainers;
exports.borderRightRadius = borderRightRadius;
exports.borderTopLeftRadius = borderTopLeftRadius;
exports.borderTopRadius = borderTopRadius;
exports.borderTopRightRadius = borderTopRightRadius;

var _variables = require("../variables/_variables");

function borderRadius() {
  let radius = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '4px';

  if (_variables.enableRounded) {
    return `border-radius: ${radius};`;
  }

  return '';
}

function borderRadiusContainers() {
  let radius = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '4px';

  if (_variables.enableRoundedContainers) {
    return `border-radius: ${radius};`;
  }

  return '';
}

function borderTopLeftRadius(radius) {
  if (_variables.enableRounded) {
    return `border-top-left-radius: ${radius};`;
  }

  return '';
}

function borderTopRightRadius(radius) {
  if (_variables.enableRounded) {
    return `border-top-right-radius: ${radius};`;
  }

  return '';
}

function borderBottomRightRadius(radius) {
  if (_variables.enableRounded) {
    return `border-bottom-right-radius: ${radius};`;
  }

  return '';
}

function borderBottomLeftRadius(radius) {
  if (_variables.enableRounded) {
    return `border-bottom-left-radius: ${radius};`;
  }

  return '';
}

function borderTopRadius(radius) {
  return `
    ${borderTopRightRadius(radius)}
    ${borderTopLeftRadius(radius)}
  `;
}

function borderRightRadius(radius) {
  return `
    ${borderBottomRightRadius(radius)}
    ${borderTopRightRadius(radius)}
  `;
}

function borderBottomRadius(radius) {
  return `
    ${borderBottomRightRadius(radius)}
    ${borderBottomLeftRadius(radius)}
  `;
}

function borderLeftRadius(radius) {
  return `
    ${borderBottomLeftRadius(radius)}
    ${borderTopLeftRadius(radius)}
  `;
}
