// @ux/stylevars/lib/mixins/gradients.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.gradientBg = gradientBg;
exports.gradientDirectional = gradientDirectional;
exports.gradientRadial = gradientRadial;
exports.gradientStriped = gradientStriped;
exports.gradientXThreeColors = gradientXThreeColors;
exports.gradientY = gradientY;
exports.gradientYThreeColors = gradientYThreeColors;

var _polished = require("polished");

var _variables = require("../variables/_variables");

/* eslint-disable max-len */
function gradientBg(color) {
  if (_variables.enableGradients) {
    return `background: ${color} linear-gradient(180deg, ${(0, _polished.mix)(0.15, _variables.bodyBg, color)} repeat-x`;
  }

  return `background-color: ${color}`;
}

function gradientY() {
  let startColor = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '#555';
  let endColor = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '#333';
  let startPercent = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : '0%';
  let endPercent = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : '100%';
  return `
    background-image: linear-gradient(to bottom, ${startColor}, ${startPercent}, ${endColor}, ${endPercent});
    backgrond-repeat: repeat-x;
  `;
}

function gradientDirectional() {
  let startColor = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '#555';
  let endColor = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '#333';
  let deg = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : '45deg';
  return `
    background-image: linear-gradient(${deg}, ${startColor}, ${endColor});
    background-repeat: repeat-x;
  `;
}

function gradientXThreeColors() {
  let startColor = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '#00b3ee';
  let midColor = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '#7a43b6';
  let colorStop = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : '50%';
  let endColor = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : '#c3325f';
  return `
    background-image: linear-gradient(to right, ${startColor}, ${midColor} ${colorStop}, ${endColor});
    background-repeat: no-repeat;
  `;
}

function gradientYThreeColors() {
  let startColor = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '#00b3ee';
  let midColor = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '#7a43b6';
  let colorStop = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : '50%';
  let endColor = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : '#c3325f';
  return `
    background-image: linear-gradient(${startColor}, ${midColor} ${colorStop}, ${endColor});
    background-repeat: no-repeat;
  `;
}

function gradientRadial() {
  let innerColor = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '#555';
  let outerColor = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '#333';
  return `
    background-image: radial-gradient(circle, ${innerColor}, ${outerColor});
    background-repeat: no-repeat;
  `;
}

function gradientStriped() {
  let color = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : (0, _polished.rgba)(255, 255, 255, 0.15);
  let angle = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '45deg';
  // eslint-disable-next-line max-len
  return `background-image: linear-gradient(${angle}, ${color} 25%, transparent 25%, transparent 50%, ${color} 50%, ${color} 75%, transparent 75%, transparent);`;
}
