// @ux/stylevars/lib/mixins/types.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.barHeader = barHeader;
exports.fontBase = fontBase;
exports.fontMarketing = fontMarketing;
exports.fontPrimaryBold = fontPrimaryBold;
exports.fontSage = fontSage;
exports.fontSageBold = fontSageBold;
exports.heading = heading;
exports.menuTitle = menuTitle;

var _markets = require("../functions/markets");

var _palette = _interopRequireDefault(require("../functions/palette"));

// Used to create a bottom margin that is based on the heading's font-size and base our line-height.
function heading(size, height) {
  return `
    font-size: ${size};
    line-height: ${height};
  `;
} // Used in Dropdowns, Growls, Tippers


function barHeader() {
  return `
    position: relative;
    color: ${(0, _palette.default)('black')};
  `;
}

function fontBase() {
  return `font-family: ${(0, _markets.marketFont)('base')};`;
}

function fontPrimaryBold() {
  return `
    font-family: ${(0, _markets.marketFont)('primary-bold', 'family')};
    font-weight: ${(0, _markets.marketFont)('primary-bold', 'weight')};
  `;
} // gd-sage only


function fontMarketing() {
  return `
    font-family: ${(0, _markets.marketFont)('marketing', 'family')};
    font-weight: ${(0, _markets.marketFont)('marketing', 'weight')} !important;
  `;
} // GD-Sage


function fontSage() {
  return `
    font-family: ${(0, _markets.marketFont)('sage', 'family')};
    font-weight: ${(0, _markets.marketFont)('sage', 'weight')};
  `;
}

function fontSageBold() {
  return `
    font-family: ${(0, _markets.marketFont)('sage-bold', 'family')};
    font-weight: ${(0, _markets.marketFont)('sage-bold', 'weight')};
  `;
} // Used for .headline-primary and form things.


function menuTitle() {
  return fontPrimaryBold();
}
