// @ux/text-entry/examples/multiline.stories.js
import TextEntry from '../src';
import React, { useState } from 'react';

export default function Multiline() {
  const [value, setValue] = useState({
    normal: 'You can type multiple lines here',
    vertical: 'You can resize vertically here here',
    placeholder: '',
    floating: '',
    both: 'You can resize in both directions here if you want. This has size={ 4 }'
  });

  return (
    <>
      <fieldset>
        <TextEntry
          type='text'
          multiline
          id='default'
          onChange={ (normal) => setValue({ ...value, normal }) }
          value={ value.normal }/>
      </fieldset>

      <fieldset>
        <TextEntry
          type='text'
          multiline
          resize='vertical'
          id='default'
          onChange={ (vertical) => setValue({ ...value, vertical }) }
          value={ value.vertical }/>
      </fieldset>

      <fieldset>
        <TextEntry
          type='text'
          multiline
          id='default'
          placeholder='This is a placeholder'
          onChange={ (placeholder) => setValue({ ...value, placeholder }) }
          value={ value.placeholder }/>
      </fieldset>

      <fieldset>
        <TextEntry
          type='text'
          multiline
          id='default'
          label='This is a floating label'
          onChange={ (label) => setValue({ ...value, label }) }
          value={ value.label }/>
      </fieldset>

      <fieldset>
        <TextEntry
          type='text'
          multiline
          resize
          size={ 4 }
          id='default'
          onChange={ (both) => setValue({ ...value, both }) }
          value={ value.both }/>
      </fieldset>
    </>
  );
}
